// @univerjs/preset-sheets-core@0.11.0/lib/index.js
import { UniverDocsPlugin as P } from "@univerjs/docs";
export * from "@univerjs/docs";
import { UniverDocsUIPlugin as v } from "@univerjs/docs-ui";
export * from "@univerjs/docs-ui";
import { UniverFormulaEnginePlugin as C } from "@univerjs/engine-formula";
export * from "@univerjs/engine-formula";
import { UniverRenderEnginePlugin as b } from "@univerjs/engine-render";
export * from "@univerjs/engine-render";
import { UniverNetworkPlugin as R } from "@univerjs/network";
export * from "@univerjs/network";
import { UniverRPCMainThreadPlugin as k } from "@univerjs/rpc";
export * from "@univerjs/rpc";
import { UniverSheetsPlugin as w } from "@univerjs/sheets";
export * from "@univerjs/sheets";
import { UniverSheetsFormulaPlugin as y } from "@univerjs/sheets-formula";
export * from "@univerjs/sheets-formula";
import { UniverSheetsFormulaUIPlugin as M } from "@univerjs/sheets-formula-ui";
export * from "@univerjs/sheets-formula-ui";
import { UniverSheetsNumfmtPlugin as A } from "@univerjs/sheets-numfmt";
export * from "@univerjs/sheets-numfmt";
import { UniverSheetsNumfmtUIPlugin as T } from "@univerjs/sheets-numfmt-ui";
export * from "@univerjs/sheets-numfmt-ui";
import { UniverSheetsUIPlugin as E } from "@univerjs/sheets-ui";
export * from "@univerjs/sheets-ui";
import { UniverUIPlugin as I } from "@univerjs/ui";
export * from "@univerjs/ui";
import "@univerjs/network/lib/facade";
import "@univerjs/sheets/lib/facade";
import "@univerjs/ui/lib/facade";
import "@univerjs/docs-ui/lib/facade";
import "@univerjs/sheets-ui/lib/facade";
import "@univerjs/engine-formula/lib/facade";
import "@univerjs/sheets-formula/lib/facade";
import "@univerjs/sheets-numfmt/lib/facade";
import "@univerjs/sheets-formula-ui/lib/facade";
function rr(m = {}) {
  var n;
  const {
    container: l = "app",
    workerURL: t,
    header: p,
    footer: u,
    toolbar: a,
    ribbonType: c,
    formulaBar: g,
    menu: f,
    contextMenu: S,
    disableAutoFocus: d,
    docs: i,
    sheets: r,
    formula: o,
    disableTextFormatAlert: U,
    disableTextFormatMark: x
  } = m, e = !!t;
  return {
    plugins: [
      R,
      [P, {
        hasScroll: i == null ? void 0 : i.hasScroll
      }],
      b,
      [I, {
        container: l,
        header: p,
        toolbar: a,
        ribbonType: c,
        menu: f,
        contextMenu: S,
        disableAutoFocus: d
      }],
      v,
      e ? [k, { workerURL: t }] : null,
      [C, {
        notExecuteFormula: e,
        function: o == null ? void 0 : o.function
      }],
      [w, {
        notExecuteFormula: e,
        onlyRegisterFormulaRelatedMutations: !1,
        isRowStylePrecedeColumnStyle: r == null ? void 0 : r.isRowStylePrecedeColumnStyle,
        autoHeightForMergedCells: r == null ? void 0 : r.autoHeightForMergedCells,
        freezeSync: r == null ? void 0 : r.freezeSync
      }],
      [E, {
        formulaBar: g,
        footer: u,
        maxAutoHeightCount: r == null ? void 0 : r.maxAutoHeightCount,
        clipboardConfig: r == null ? void 0 : r.clipboardConfig,
        scrollConfig: r == null ? void 0 : r.scrollConfig,
        protectedRangeShadow: (n = r == null ? void 0 : r.protectedRangeShadow) != null ? n : !0,
        protectedRangeUserSelector: r == null ? void 0 : r.protectedRangeUserSelector,
        disableForceStringAlert: r == null ? void 0 : r.disableForceStringAlert,
        disableForceStringMark: r == null ? void 0 : r.disableForceStringMark
      }],
      [A, {
        disableTextFormatAlert: U,
        disableTextFormatMark: x
      }],
      T,
      [y, {
        notExecuteFormula: e,
        description: o == null ? void 0 : o.description,
        initialFormulaComputing: o == null ? void 0 : o.initialFormulaComputing
      }],
      [M, {
        functionScreenTips: o == null ? void 0 : o.functionScreenTips
      }]
    ].filter((F) => !!F)
  };
}
export {
  rr as UniverSheetsCorePreset
};
