// @univerjs/sheets-formula-ui/lib/es/index.js
var Yr = Object.defineProperty;
var Gr = (e, t, n) => t in e ? Yr(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n;
var j = (e, t, n) => Gr(e, typeof t != "symbol" ? t + "" : t, n);
import { CommandType as Ne, ICommandService as me, createIdentifier as Zr, IContextService as cn, IUniverInstanceService as ne, Rectangle as er, DOCS_NORMAL_EDITOR_UNIT_ID_KEY as an, DOCS_FORMULA_BAR_EDITOR_UNIT_ID_KEY as tr, DEFAULT_EMPTY_DOCUMENT_VALUE as nr, isRealNum as zr, CellValueType as xn, getCellValueType as Xr, Inject as q, Disposable as ut, ObjectMatrix as Fe, Range as Qr, Tools as ln, LocaleService as dt, isICellData as Jr, isFormulaString as Ae, isFormulaId as at, generateRandomId as qt, Direction as de, Injector as Rt, UniverInstanceType as H, ThemeService as It, ILogService as eo, toDisposable as rr, ColorKit as un, RxDisposable as to, InterceptorEffectEnum as no, FOCUSING_DOC as ro, FOCUSING_UNIVER_EDITOR as oo, DisposableCollection as qe, RANGE_TYPE as ve, getBodySlice as Mn, EDITOR_ACTIVATED as wn, createInternalEditorID as so, BuildTextUtils as io, IConfigService as or, RichTextBuilder as co, DependentOn as ao, Plugin as lo, merge as uo, registerDependencies as ho, touchDependencies as fo } from "@univerjs/core";
import { SheetPasteCommand as go, PREDEFINED_HOOK_NAME as it, IEditorBridgeService as dn, SetCellEditVisibleOperation as sr, HoverManagerService as mo, CellAlertManagerService as po, CellAlertType as So, IAutoFillService as vo, APPLY_TYPE as Co, DATA_TYPE as kn, ISheetClipboardService as _o, COPY_TYPE as ir, SheetSkeletonManagerService as jt, attachSelectionWithCoord as tn, SelectionControl as cr, SELECTION_SHAPE_DEPTH as Ro, useActiveWorkbook as Io, getCurrentRangeDisable$ as Ke, PASTE_SPECIAL_MENU_ID as Eo, whenFormulaEditorActivated as Et, whenSheetEditorFocused as bo, SheetsUIPart as yo, BaseSelectionRenderService as To, getCoordByOffset as An, checkInHeaderRanges as Fn, getAllSelection as No, genNormalSelectionStyle as ar, getSheetObject as Oo, MoveSelectionCommand as Dn, JumpOver as Ln, ExpandSelectionCommand as $n, EMBEDDING_FORMULA_EDITOR as xo, IMarkSelectionService as Mo, RANGE_SELECTOR_COMPONENT_KEY as wo, EMBEDDING_FORMULA_EDITOR_COMPONENT_KEY as ko } from "@univerjs/sheets-ui";
import { sequenceNodeType as Q, serializeRange as Ce, FormulaDataModel as bt, LexerTreeBuilder as _e, ErrorType as ue, extractFormulaError as lr, SetFormulaCalculationResultMutation as Ao, SetArrayFormulaDataMutation as Fo, SetFormulaCalculationStopMutation as Do, FunctionType as ur, matchToken as je, deserializeRangeWithSheetWithCache as Lo, matchRefDrawToken as $o, isFormulaLexerToken as Po, deserializeRangeWithSheet as ht, serializeRangeToRefString as Uo, serializeRangeWithSheet as Ct, serializeRangeWithSpreadsheet as Vo, generateStringWithSequence as Wo, operatorToken as dr, UniverFormulaEnginePlugin as Ho } from "@univerjs/engine-formula";
import { Subject as mt, debounceTime as hn, combineLatestWith as Bo, map as hr, switchMap as qo, of as Pn, Observable as jo, BehaviorSubject as fr, throttleTime as Ko, filter as gr, distinctUntilChanged as Yo, merge as Go } from "rxjs";
import { IEditorService as Ye, DocSelectionRenderService as mr, ReplaceTextRunsCommand as Un, MoveSelectionOperation as Zo, MoveCursorOperation as zo, useKeyboardEvent as Xo, useResize as Qo, DocBackScrollRenderController as Jo, RichTextEditor as es } from "@univerjs/docs-ui";
import { DeviceInputEventType as Oe, IRenderManagerService as xe, ScrollTimerType as Xt, SHEET_VIEWPORT_KEY as Vn, Vector2 as Wn } from "@univerjs/engine-render";
import { SheetsSelectionsService as fn, getSheetCommandTarget as pr, getCellAtRowCol as ts, SetSelectionsOperation as gn, SheetInterceptorService as mn, ReorderRangeCommand as ns, SetRangeValuesMutation as Ft, SetRangeValuesUndoMutationFactory as rs, BEFORE_CELL_EDIT as os, SetWorksheetRowAutoHeightMutation as ss, INTERCEPTOR_POINT as is, WorksheetSetCellValuePermission as Ge, WorksheetEditPermission as Ze, RangeProtectionPermissionEditPoint as ze, WorkbookEditablePermission as Xe, IRefSelectionsService as Kt, SelectionMoveType as Qt, convertSelectionDataToRange as cs, setEndForRange as as, REF_SELECTIONS_ENABLED as Hn, SetWorksheetActiveOperation as Bn } from "@univerjs/sheets";
import { InsertFunctionCommand as ls, TriggerCalculationController as us, IDescriptionService as Yt, QuickSumCommand as ds, UniverSheetsFormulaPlugin as hs } from "@univerjs/sheets-formula";
import { ISidebarService as pn, IZenZoneService as fs, useDependency as T, useObservable as Se, ProgressBar as gs, MenuItemType as Qe, getMenuHiddenObservable as ft, IClipboardInterfaceService as qn, RibbonFormulasGroup as jn, KeyCode as L, MetaKeys as P, IMenuManagerService as ms, IShortcutService as Gt, IUIPartsService as Sr, ComponentManager as vr, connectInjector as Cr, useEvent as te, RectPopup as nn, IContextMenuService as ps, useUpdateEffect as Ss, BuiltInUIPart as vs } from "@univerjs/ui";
import { jsx as M, jsxs as W, Fragment as Cs } from "react/jsx-runtime";
import { useCallback as Dt, useState as V, useRef as z, createElement as Je, forwardRef as Le, useEffect as $, useMemo as ie, useLayoutEffect as Lt, useImperativeHandle as _s } from "react";
import { clsx as re, scrollbarClassName as lt, borderLeftClassName as Rs, Select as Is, Input as _r, borderClassName as Sn, Button as ct, borderTopClassName as Es, Tooltip as bs, Dialog as ys } from "@univerjs/design";
import { DocSelectionManagerService as Rr } from "@univerjs/docs";
import { debounceTime as Ts } from "rxjs/operators";
const vn = {
  id: "sheet.command.paste-formula",
  type: Ne.COMMAND,
  handler: async (e) => e.get(me).executeCommand(go.id, {
    value: it.SPECIAL_PASTE_FORMULA
  })
}, gt = {
  id: "formula-ui.operation.select-editor-formula",
  type: Ne.OPERATION,
  handler: (e, t) => !0
};
var Ns = Object.getOwnPropertyDescriptor, Os = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Ns(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, xs = (e, t) => (n, o) => t(n, o, e);
const Ms = "FORMULA_PROMPT_ACTIVATED", Zt = Zr("formula-ui.prompt-service");
let rn = class {
  constructor(e) {
    j(this, "_search$", new mt());
    j(this, "_help$", new mt());
    j(this, "_navigate$", new mt());
    j(this, "_accept$", new mt());
    j(this, "_acceptFormulaName$", new mt());
    j(this, "search$", this._search$.asObservable());
    j(this, "help$", this._help$.asObservable());
    j(this, "navigate$", this._navigate$.asObservable());
    j(this, "accept$", this._accept$.asObservable());
    j(this, "acceptFormulaName$", this._acceptFormulaName$.asObservable());
    j(this, "_searching", !1);
    j(this, "_helping", !1);
    j(this, "_sequenceNodes", []);
    j(this, "_isLockedOnSelectionChangeRefString", !1);
    j(this, "_isLockedOnSelectionInsertRefString", !1);
    this._contextService = e;
  }
  dispose() {
    this._search$.complete(), this._help$.complete(), this._navigate$.complete(), this._accept$.complete(), this._acceptFormulaName$.complete(), this._sequenceNodes = [];
  }
  search(e) {
    this._contextService.setContextValue(Ms, e.visible), this._searching = e.visible, this._search$.next(e);
  }
  isSearching() {
    return this._searching;
  }
  help(e) {
    this._helping = e.visible, this._help$.next(e);
  }
  isHelping() {
    return this._helping;
  }
  navigate(e) {
    this._navigate$.next(e);
  }
  accept(e) {
    this._accept$.next(e);
  }
  acceptFormulaName(e) {
    this._acceptFormulaName$.next(e);
  }
  getSequenceNodes() {
    return [...this._sequenceNodes];
  }
  setSequenceNodes(e) {
    this._sequenceNodes = e;
  }
  clearSequenceNodes() {
    this._sequenceNodes = [];
  }
  getCurrentSequenceNode(e) {
    return this._sequenceNodes[this.getCurrentSequenceNodeIndex(e)];
  }
  getCurrentSequenceNodeByIndex(e) {
    return this._sequenceNodes[e];
  }
  /**
   * Query the text coordinates in the sequenceNodes and determine the actual insertion index.
   * @param strIndex
   */
  getCurrentSequenceNodeIndex(e) {
    let t = 0;
    const n = this._sequenceNodes[0];
    for (let o = 0, r = this._sequenceNodes.length; o < r; o++) {
      const s = this._sequenceNodes[o];
      if (typeof s == "string")
        t++;
      else {
        const { endIndex: i } = s;
        t = i;
      }
      if (e <= t)
        return typeof n == "string" && e !== 0 ? o + 1 : o;
    }
    return this._sequenceNodes.length;
  }
  /**
   * Synchronize the reference text based on the changes of the selection.
   * @param nodeIndex
   * @param refString
   */
  updateSequenceRef(e, t) {
    const n = this._sequenceNodes[e];
    if (typeof n == "string" || n.nodeType !== Q.REFERENCE)
      return;
    const o = t.length - n.token.length, r = { ...n };
    r.token = t, r.endIndex += o, this._sequenceNodes[e] = r;
    for (let s = e + 1, i = this._sequenceNodes.length; s < i; s++) {
      const c = this._sequenceNodes[s];
      if (typeof c == "string")
        continue;
      const l = { ...c };
      l.startIndex += o, l.endIndex += o, this._sequenceNodes[s] = l;
    }
  }
  /**
   * When the cursor is on the right side of a formula token,
   * you can add reference text to the formula by drawing a selection.
   * @param index
   * @param refString
   */
  insertSequenceRef(e, t) {
    const n = t.length, o = this.getCurrentSequenceNodeIndex(e);
    this._sequenceNodes.splice(o, 0, {
      token: t,
      startIndex: e,
      endIndex: e + n - 1,
      nodeType: Q.REFERENCE
    });
    for (let r = o + 1, s = this._sequenceNodes.length; r < s; r++) {
      const i = this._sequenceNodes[r];
      if (typeof i == "string")
        continue;
      const c = { ...i };
      c.startIndex += n, c.endIndex += n, this._sequenceNodes[r] = c;
    }
  }
  /**
   * Insert a string at the cursor position in the text corresponding to the sequenceNodes.
   * @param index
   * @param content
   */
  insertSequenceString(e, t) {
    const n = this.getCurrentSequenceNodeIndex(e), o = t.split("");
    this._sequenceNodes.splice(n, 0, ...o);
    const r = o.length;
    for (let s = n + r, i = this._sequenceNodes.length; s < i; s++) {
      const c = this._sequenceNodes[s];
      if (typeof c == "string")
        continue;
      const l = { ...c };
      l.startIndex += r, l.endIndex += r, this._sequenceNodes[s] = l;
    }
  }
  enableLockedSelectionChange() {
    this._isLockedOnSelectionChangeRefString = !0;
  }
  disableLockedSelectionChange() {
    this._isLockedOnSelectionChangeRefString = !1;
  }
  isLockedSelectionChange() {
    return this._isLockedOnSelectionChangeRefString;
  }
  enableLockedSelectionInsert() {
    this._isLockedOnSelectionInsertRefString = !0;
  }
  disableLockedSelectionInsert() {
    this._isLockedOnSelectionInsertRefString = !1;
  }
  isLockedSelectionInsert() {
    return this._isLockedOnSelectionInsertRefString;
  }
};
rn = Os([
  xs(0, cn)
], rn);
const ws = {
  id: "formula-ui.operation.help-function",
  type: Ne.OPERATION,
  handler: async (e, t) => (e.get(Zt).help(t), !0)
}, pe = {
  id: "formula-ui.operation.insert-function",
  type: Ne.OPERATION,
  // eslint-disable-next-line
  handler: async (e, t) => {
    var C, _;
    const n = e.get(fn), o = e.get(Ye), r = n.getCurrentSelections();
    if (!r || !r.length)
      return !1;
    const s = pr(e.get(ne));
    if (!s) return !1;
    const { worksheet: i, unitId: c, subUnitId: l } = s, g = i.getCellMatrix(), { value: m } = t, h = e.get(me);
    e.get(dn);
    const d = [], u = [];
    let a = null, v = 0, p = 0, S = "";
    if (r.length === 1 && (Fs(r[0].range) || Ds(r[0].range) && Yn(g, r[0].range))) {
      const { range: E, primary: f } = r[0], b = (C = f == null ? void 0 : f.actualRow) != null ? C : E.startRow, y = (_ = f == null ? void 0 : f.actualColumn) != null ? _ : E.startColumn;
      a = E, v = b, p = y;
      const A = Kn(g, b, y);
      A && (S = Ce(A));
    } else
      r.some((E) => {
        var y, A;
        const { range: f, primary: b } = E;
        if (Yn(g, f)) {
          const F = (y = b == null ? void 0 : b.actualRow) != null ? y : f.startRow, N = (A = b == null ? void 0 : b.actualColumn) != null ? A : f.startColumn, O = Kn(g, F, N);
          if (!O)
            return a = f, v = F, p = N, !0;
          const k = Ce(O), I = `=${m}(${k})`;
          d.push({
            range: f,
            primary: {
              row: F,
              column: N
            },
            formula: I
          });
        } else {
          const { startRow: F, startColumn: N, endRow: O, endColumn: k } = f;
          if (F === O) {
            const I = Ls(g, F, k, i.getColumnCount() - 1), x = I === k ? k - 1 : k, R = Ce({
              startRow: F,
              endRow: O,
              startColumn: N,
              endColumn: x
            }), w = `=${m}(${R})`;
            u.push({
              range: f,
              primary: {
                row: F,
                column: I
              },
              formula: w
            });
          } else {
            let I = -1;
            for (let R = N; R <= k; R++) {
              const w = $s(g, R, O, i.getRowCount() - 1);
              I = Math.max(I, w);
            }
            const x = I === O ? O - 1 : O;
            for (let R = N; R <= k; R++) {
              const w = Ce({
                startRow: F,
                endRow: x,
                startColumn: R,
                endColumn: R
              }), D = `=${m}(${w})`;
              u.push({
                range: f,
                primary: {
                  row: I,
                  column: R
                },
                formula: D
              });
            }
          }
        }
        return !1;
      });
    if (a) {
      const E = ts(v, p, i), f = {
        range: er.clone(a),
        primary: {
          startRow: E.startRow,
          startColumn: E.startColumn,
          endRow: E.endRow,
          endColumn: E.endColumn,
          actualRow: v,
          actualColumn: p,
          isMerged: E.isMerged,
          isMergedMainCell: E.startRow === v && E.startColumn === p
        }
      }, b = {
        unitId: c,
        subUnitId: l,
        selections: [f]
      };
      await h.executeCommand(gn.id, b);
      const y = o.getEditor(an), A = o.getEditor(tr);
      h.syncExecuteCommand(sr.id, {
        visible: !0,
        unitId: c,
        eventType: Oe.Dblclick
      });
      const F = `=${m}(${S}`;
      y == null || y.replaceText(F), A == null || A.replaceText(F, !1);
    }
    return d.length === 0 && u.length === 0 ? !1 : h.executeCommand(ls.id, {
      list: d,
      listOfRangeHasNumber: u
    });
  }
};
function Kn(e, t, n) {
  const o = ks(e, t, n);
  if (o !== t)
    return {
      startRow: o,
      endRow: t - 1,
      startColumn: n,
      endColumn: n
    };
  const r = As(e, t, n);
  return r !== n ? {
    startRow: t,
    endRow: t,
    startColumn: r,
    endColumn: n - 1
  } : null;
}
function ks(e, t, n) {
  let o = !1;
  if (t === 0) return t;
  for (let r = t - 1; r >= 0; r--) {
    const s = e.getValue(r, n);
    if (_t(s) && !o) {
      if (r === 0) return 0;
      o = !0;
    } else {
      if (o && !_t(s))
        return r + 1;
      if (o && r === 0)
        return 0;
    }
  }
  return t;
}
function As(e, t, n) {
  let o = !1;
  if (n === 0) return n;
  for (let r = n - 1; r >= 0; r--) {
    const s = e.getValue(t, r);
    if (_t(s) && !o) {
      if (r === 0) return 0;
      o = !0;
    } else {
      if (o && !_t(s))
        return r + 1;
      if (o && r === 0)
        return 0;
    }
  }
  return n;
}
function _t(e) {
  if (e != null && e.p) {
    const t = e == null ? void 0 : e.p.body;
    if (t == null)
      return !1;
    const n = t.dataStream, r = n.substring(n.length - 2, n.length) === nr ? n.substring(0, n.length - 2) : n;
    return zr(r);
  }
  return e && (e.t === xn.NUMBER || Xr(e) === xn.NUMBER);
}
function Fs(e) {
  return e.startRow === e.endRow && e.startColumn === e.endColumn;
}
function Ds(e) {
  return e.startRow !== e.endRow && e.startColumn !== e.endColumn;
}
function Yn(e, t) {
  for (let n = t.startRow; n <= t.endRow; n++)
    for (let o = t.startColumn; o <= t.endColumn; o++)
      if (_t(e.getValue(n, o)))
        return !1;
  return !0;
}
function Ls(e, t, n, o) {
  for (let r = n; r <= o; r++)
    if (!e.getValue(t, r))
      return r;
  return o;
}
function $s(e, t, n, o) {
  for (let r = n; r <= o; r++)
    if (!e.getValue(r, t))
      return r;
  return o;
}
const Ir = "SHEET_FORMULA_UI_PLUGIN", Er = `${Ir}_MORE_FUNCTIONS_COMPONENT`, Cn = {
  id: "formula-ui.operation.more-functions",
  type: Ne.OPERATION,
  handler: async (e) => (e.get(pn).open({
    header: { title: "formula.insert.tooltip" },
    children: { label: Er }
  }), !0)
}, br = {
  id: "formula-ui.operation.change-ref-to-absolute",
  type: Ne.OPERATION,
  handler: async (e) => !0
}, Ps = {
  id: "formula-ui.operation.search-function",
  type: Ne.OPERATION,
  handler: async (e, t) => (e.get(Zt).search(t), !0)
};
var Us = Object.getOwnPropertyDescriptor, Vs = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Us(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, wt = (e, t) => (n, o) => t(n, o, e);
let $t = class extends ut {
  constructor(e, t, n, o) {
    super(), this._sheetInterceptorService = e, this._univerInstanceService = t, this._formulaDataModel = n, this._lexerTreeBuilder = o, this._initialize();
  }
  _initialize() {
    this.disposeWithMe(this._sheetInterceptorService.interceptCommand({
      getMutations: (e) => e.id === ns.id ? this._reorderFormula(e.params) : {
        redos: [],
        undos: []
      }
    }));
  }
  _reorderFormula(e) {
    const t = [], n = [], { unitId: o, subUnitId: r, range: s, order: i } = e, c = this._univerInstanceService.getUniverSheetInstance(o), l = c == null ? void 0 : c.getSheetBySheetId(r);
    if (!l)
      return {
        redos: t,
        undos: n
      };
    const g = l.getCellMatrix(), m = new Fe(), h = new Fe();
    let d = !1;
    return Qr.foreach(s, (u, a) => {
      let v = u;
      i.hasOwnProperty(u) && (v = i[u]);
      const p = g.getValue(v, a);
      if (p != null && p.f || p != null && p.si) {
        d = !0;
        const S = this._formulaDataModel.getFormulaStringByCell(v, a, r, o), C = this._lexerTreeBuilder.moveFormulaRefOffset(
          S,
          0,
          u - v
        ), _ = ln.deepClone(p);
        _.f = C, _.si = null, m.setValue(u, a, _);
      } else
        m.setValue(u, a, p);
      h.setValue(u, a, g.getValue(u, a));
    }), d ? (t.push({
      id: Ft.id,
      params: {
        unitId: o,
        subUnitId: r,
        cellValue: m.getMatrix()
      }
    }), n.push({
      id: Ft.id,
      params: {
        unitId: o,
        subUnitId: r,
        cellValue: h.getMatrix()
      }
    }), {
      redos: t,
      undos: n
    }) : {
      redos: t,
      undos: n
    };
  }
};
$t = Vs([
  wt(0, q(mn)),
  wt(1, q(ne)),
  wt(2, q(bt)),
  wt(3, q(_e))
], $t);
const yr = "sheets-formula-ui.base.config", Gn = {};
var Ws = Object.getOwnPropertyDescriptor, Hs = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Ws(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, pt = (e, t) => (n, o) => t(n, o, e);
const Jt = "SHEET_FORMULA_ALERT", Bs = {
  [ue.DIV_BY_ZERO]: "divByZero",
  [ue.NAME]: "name",
  [ue.VALUE]: "value",
  [ue.NUM]: "num",
  [ue.NA]: "na",
  [ue.CYCLE]: "cycle",
  [ue.REF]: "ref",
  [ue.SPILL]: "spill",
  [ue.CALC]: "calc",
  [ue.ERROR]: "error",
  [ue.CONNECT]: "connect",
  [ue.NULL]: "null"
};
let on = class extends ut {
  constructor(e, t, n, o, r, s) {
    super(), this._context = e, this._hoverManagerService = t, this._cellAlertManagerService = n, this._localeService = o, this._formulaDataModel = r, this._zenZoneService = s, this._init();
  }
  _init() {
    this._initCellAlertPopup(), this._initZenService();
  }
  _initCellAlertPopup() {
    this.disposeWithMe(this._hoverManagerService.currentCell$.pipe(hn(100)).subscribe((e) => {
      var t, n, o, r, s;
      if (e) {
        const c = this._context.unit.getActiveSheet();
        if (!c) return this._hideAlert();
        const l = c.getCell(e.location.row, e.location.col), g = (r = (o = (n = (t = this._formulaDataModel.getArrayFormulaCellData()) == null ? void 0 : t[e.location.unitId]) == null ? void 0 : n[e.location.subUnitId]) == null ? void 0 : o[e.location.row]) == null ? void 0 : r[e.location.col];
        if (Jr(l)) {
          const m = lr(l, !!g);
          if (!m) {
            this._hideAlert();
            return;
          }
          const h = this._cellAlertManagerService.currentAlert.get(Jt), d = (s = h == null ? void 0 : h.alert) == null ? void 0 : s.location;
          if (d && d.row === e.location.row && d.col === e.location.col && d.subUnitId === e.location.subUnitId && d.unitId === e.location.unitId) {
            this._hideAlert();
            return;
          }
          this._cellAlertManagerService.showAlert({
            type: So.ERROR,
            title: this._localeService.t("formula.error.title"),
            message: this._localeService.t(`formula.error.${Bs[m]}`),
            location: e.location,
            width: 200,
            height: 74,
            key: Jt
          });
          return;
        }
      }
      this._hideAlert();
    }));
  }
  _initZenService() {
    this.disposeWithMe(this._zenZoneService.visible$.subscribe((e) => {
      e && this._hideAlert();
    }));
  }
  _hideAlert() {
    this._cellAlertManagerService.removeAlert(Jt);
  }
};
on = Hs([
  pt(1, q(mo)),
  pt(2, q(po)),
  pt(3, q(dt)),
  pt(4, q(bt)),
  pt(5, fs)
], on);
var qs = Object.getOwnPropertyDescriptor, js = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? qs(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, Zn = (e, t) => (n, o) => t(n, o, e);
let Pt = class extends ut {
  constructor(e, t) {
    super(), this._autoFillService = e, this._lexerTreeBuilder = t, this._registerAutoFill();
  }
  _registerAutoFill() {
    const e = {
      type: kn.FORMULA,
      priority: 1001,
      match: (t) => Ae(t == null ? void 0 : t.f) || at(t == null ? void 0 : t.si),
      isContinue: (t, n) => t.type === kn.FORMULA,
      applyFunctions: {
        [Co.COPY]: (t, n, o, r, s) => {
          const { data: i, index: c } = t;
          return this._fillCopyFormula(i, n, o, c, r, s);
        }
      }
    };
    this._autoFillService.registerRule(e);
  }
  _fillCopyFormula(e, t, n, o, r, s) {
    var g, m;
    const i = Ys(r), c = [], l = /* @__PURE__ */ new Map();
    for (let h = 1; h <= t; h++) {
      const d = (h - 1) % e.length, u = o[d], a = ln.deepClone(e[d]);
      if (a) {
        const v = ((g = e[d]) == null ? void 0 : g.f) || "", p = ((m = e[d]) == null ? void 0 : m.si) || "", S = Ae(v);
        if (at(p))
          a.si = p, a.f = null, a.v = null, a.p = null, a.t = null, c.push(a);
        else if (S) {
          let _ = l.get(d);
          if (_)
            a.si = _, a.f = null, a.v = null, a.p = null, a.t = null;
          else {
            _ = qt(6), l.set(d, _);
            const { offsetX: E, offsetY: f } = Ks(i, t, n, s, u), b = this._lexerTreeBuilder.moveFormulaRefOffset(
              v,
              E,
              f
            );
            a.si = _, a.f = b, a.v = null, a.p = null, a.t = null;
          }
          c.push(a);
        }
      }
    }
    return c;
  }
};
Pt = js([
  Zn(0, vo),
  Zn(1, q(_e))
], Pt);
function Ks(e, t, n, o, r) {
  const { source: s, target: i } = o, { rows: c } = i, { rows: l } = s;
  let g = 0, m = 0;
  switch (n) {
    case de.UP:
      m = c[r] - l[r];
      break;
    case de.RIGHT:
      g = e;
      break;
    case de.DOWN:
      m = c[r] - l[r];
      break;
    case de.LEFT:
      g = -e * t;
      break;
  }
  return { offsetX: g, offsetY: m };
}
function Ys(e) {
  let t = 0;
  for (const n in e)
    e[n].forEach((o) => {
      t += o.data.length;
    });
  return t;
}
var Gs = Object.getOwnPropertyDescriptor, Zs = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Gs(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, St = (e, t) => (n, o) => t(n, o, e);
const zs = "default-paste-formula";
let Ut = class extends ut {
  constructor(e, t, n, o, r) {
    super(), this._currentUniverSheet = e, this._lexerTreeBuilder = t, this._sheetClipboardService = n, this._injector = o, this._formulaDataModel = r, this._initialize();
  }
  _initialize() {
    this._registerClipboardHook();
  }
  _registerClipboardHook() {
    this.disposeWithMe(this._sheetClipboardService.addClipboardHook(this._pasteFormulaHook())), this.disposeWithMe(this._sheetClipboardService.addClipboardHook(this._pasteWithFormulaHook()));
  }
  _pasteFormulaHook() {
    return {
      id: it.SPECIAL_PASTE_FORMULA,
      priority: 10,
      specialPasteInfo: { label: "specialPaste.formula" },
      onPasteCells: (e, t, n, o) => this._onPasteCells(e, t, n, o, !0)
    };
  }
  _pasteWithFormulaHook() {
    return {
      id: zs,
      priority: 10,
      onPasteCells: (e, t, n, o) => this._onPasteCells(e, t, n, o, !1)
    };
  }
  _onPasteCells(e, t, n, o, r) {
    var d;
    if ([
      it.SPECIAL_PASTE_FORMAT,
      it.SPECIAL_PASTE_COL_WIDTH
    ].includes(o.pasteType))
      return {
        undos: [],
        redos: []
      };
    const i = this._currentUniverSheet.getCurrentUnitForType(H.UNIVER_SHEET), c = t.unitId || i.getUnitId(), l = t.subUnitId || ((d = i.getActiveSheet()) == null ? void 0 : d.getSheetId());
    if (!c || !l)
      return {
        undos: [],
        redos: []
      };
    const g = t.range, m = n, h = {
      copyType: o.copyType || ir.COPY,
      copyRange: e == null ? void 0 : e.range,
      pasteType: o.pasteType
    };
    return this._injector.invoke((u) => Xs(
      c,
      l,
      g,
      m,
      u,
      h,
      this._lexerTreeBuilder,
      this._formulaDataModel,
      r,
      e
    ));
  }
};
Ut = Zs([
  St(0, ne),
  St(1, q(_e)),
  St(2, _o),
  St(3, q(Rt)),
  St(4, q(bt))
], Ut);
function Xs(e, t, n, o, r, s, i, c, l = !1, g) {
  const m = [], h = [], d = Qs(e, t, n, o, s, i, c, g);
  if (!d.hasValue())
    return {
      undos: [],
      redos: []
    };
  const u = {
    unitId: e,
    subUnitId: t,
    cellValue: d.getData()
  };
  m.push({
    id: Ft.id,
    params: u
  });
  const a = rs(
    r,
    u
  );
  return h.push({
    id: Ft.id,
    params: a
  }), {
    undos: h,
    redos: m
  };
}
function Qs(e, t, n, o, r, s, i, c) {
  return c ? r.pasteType === it.SPECIAL_PASTE_VALUE ? ei(e, t, n, o, i, c) : r.pasteType === it.SPECIAL_PASTE_FORMULA ? ti(e, t, n, o, s, i, c) : ni(e, t, n, o, r.copyType, s, i, c) : Js(e, t, n, o, i);
}
function Js(e, t, n, o, r) {
  const s = new Fe(), i = r.getSheetFormulaData(e, t);
  return o.forValue((c, l, g) => {
    var u;
    const m = n.rows[c], h = n.cols[l], d = {};
    Ae(g.v) ? (d.v = null, d.f = `${g.v}`, d.si = null, d.p = null, s.setValue(m, h, d)) : (u = i == null ? void 0 : i[m]) != null && u[h] && (d.v = g.v, d.f = null, d.si = null, d.p = null, s.setValue(m, h, d));
  }), s;
}
function ei(e, t, n, o, r, s) {
  var g, m;
  const i = new Fe(), c = (m = (g = r.getArrayFormulaCellData()) == null ? void 0 : g[s.unitId]) == null ? void 0 : m[s.subUnitId], l = r.getSheetFormulaData(e, t);
  return o.forValue((h, d, u) => {
    var _, E;
    const a = s.range.rows[h % s.range.rows.length], v = s.range.cols[d % s.range.cols.length], p = n.rows[h], S = n.cols[d], C = {};
    if (Ae(u.f) || at(u.si))
      C.v = u.v, C.f = null, C.si = null, C.p = null, i.setValue(p, S, C);
    else if ((_ = c == null ? void 0 : c[a]) != null && _[v]) {
      const f = c[a][v];
      C.v = f.v, C.f = null, C.si = null, C.p = null, i.setValue(p, S, C);
    } else if ((E = l == null ? void 0 : l[p]) != null && E[S]) {
      if (C.v = u.v, C.f = null, C.si = null, C.p = null, u.p) {
        const f = Tr(u);
        f && (C.v = f);
      }
      i.setValue(p, S, C);
    }
  }), i;
}
function ti(e, t, n, o, r, s, i) {
  const c = new Fe(), l = /* @__PURE__ */ new Map();
  return o.forValue((g, m, h) => {
    const d = n.rows[g], u = n.cols[m], a = {};
    if (at(h.si)) {
      if (i.unitId !== e || i.subUnitId !== t) {
        const v = s.getFormulaStringByCell(
          i.range.rows[g % i.range.rows.length],
          i.range.cols[m % i.range.cols.length],
          i.subUnitId,
          i.unitId
        ), p = n.cols[m] - i.range.cols[m % i.range.cols.length], S = n.rows[g] - i.range.rows[g % i.range.rows.length], C = r.moveFormulaRefOffset(v || "", p, S);
        a.si = null, a.f = C;
      } else
        a.si = h.si, a.f = null;
      a.v = null, a.p = null, c.setValue(d, u, a);
    } else if (Ae(h.f)) {
      const v = `${g % i.range.rows.length}_${m % i.range.cols.length}`;
      let p = l.get(v);
      if (p)
        a.si = p, a.f = null;
      else {
        p = qt(6), l.set(v, p);
        const S = n.cols[m] - i.range.cols[m % i.range.cols.length], C = n.rows[g] - i.range.rows[g % i.range.rows.length], _ = r.moveFormulaRefOffset(h.f || "", S, C);
        a.si = p, a.f = _;
      }
      a.v = null, a.p = null, c.setValue(d, u, a);
    } else {
      if (a.v = h.v, a.f = null, a.si = null, a.p = null, h.p) {
        const v = Tr(h);
        v && (a.v = v);
      }
      c.setValue(d, u, a);
    }
  }), c;
}
function ni(e, t, n, o, r, s, i, c) {
  const l = new Fe(), g = /* @__PURE__ */ new Map(), m = i.getSheetFormulaData(e, t), h = [];
  return r === ir.CUT ? o.forValue((d, u, a) => {
    const v = n.rows[d], p = n.cols[u], S = {};
    if (at(a.si)) {
      if (Ae(a.f))
        h.push(a.si), S.f = a.f, S.si = a.si;
      else if (h.includes(a.si))
        S.f = null, S.si = a.si;
      else {
        const C = i.getFormulaStringByCell(
          c.range.rows[d % c.range.rows.length],
          c.range.cols[u % c.range.cols.length],
          c.subUnitId,
          c.unitId
        );
        S.f = C, S.si = null;
      }
      S.v = null, S.p = null, l.setValue(v, p, S);
    } else Ae(a.f) && (S.f = a.f, S.si = null, S.v = null, S.p = null, l.setValue(v, p, S));
  }) : o.forValue((d, u, a) => {
    var C;
    const v = n.rows[d], p = n.cols[u], S = {};
    if (at(a.si)) {
      if (c.unitId !== e || c.subUnitId !== t) {
        const _ = i.getFormulaStringByCell(
          c.range.rows[d % c.range.rows.length],
          c.range.cols[u % c.range.cols.length],
          c.subUnitId,
          c.unitId
        ), E = n.cols[u] - c.range.cols[u % c.range.cols.length], f = n.rows[d] - c.range.rows[d % c.range.rows.length], b = s.moveFormulaRefOffset(_ || "", E, f);
        S.si = null, S.f = b;
      } else
        S.si = a.si, S.f = null;
      S.v = null, S.p = null, l.setValue(v, p, S);
    } else if (Ae(a.f)) {
      const _ = `${d % c.range.rows.length}_${u % c.range.cols.length}`;
      let E = g.get(_);
      if (E)
        S.si = E, S.f = null;
      else {
        E = qt(6), g.set(_, E);
        const f = n.cols[u] - c.range.cols[u % c.range.cols.length], b = n.rows[d] - c.range.rows[d % c.range.rows.length], y = s.moveFormulaRefOffset(a.f || "", f, b);
        S.si = E, S.f = y;
      }
      S.v = null, S.p = null, l.setValue(v, p, S);
    } else (C = m == null ? void 0 : m[v]) != null && C[p] && (S.v = a.v, S.f = null, S.si = null, S.p = a.p, l.setValue(v, p, S));
  }), h.length > 0 && new Fe(m).forValue((d, u, a) => {
    if (!(c.range.rows.includes(d) && c.range.cols.includes(u)) && !(n.rows.includes(d) && n.cols.includes(u)) && h.includes(a == null ? void 0 : a.si)) {
      const v = i.getFormulaStringByCell(
        d,
        u,
        c.subUnitId,
        c.unitId
      );
      l.setValue(d, u, {
        f: v,
        si: null,
        v: null,
        p: null
      });
    }
  }), l;
}
function Tr(e) {
  if (e != null && e.p) {
    const t = e == null ? void 0 : e.p.body;
    if (t == null)
      return;
    const n = t.dataStream;
    return n.substring(n.length - 2, n.length) === nr ? n.substring(0, n.length - 2) : n;
  }
}
var ri = Object.getOwnPropertyDescriptor, oi = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? ri(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, He = (e, t) => (n, o) => t(n, o, e);
let Vt = class extends ut {
  constructor(t, n, o, r, s, i, c, l) {
    super();
    j(this, "_previousShape");
    j(this, "_skeleton");
    this._context = t, this._sheetInterceptorService = n, this._formulaDataModel = o, this._themeService = r, this._renderManagerService = s, this._sheetSkeletonManagerService = i, this._commandService = c, this._logService = l, this._initSkeletonChangeListener(), this._initInterceptorEditorStart(), this._commandExecutedListener();
  }
  _initSkeletonChangeListener() {
    this.disposeWithMe(
      this._sheetSkeletonManagerService.currentSkeleton$.subscribe((t) => {
        var n, o;
        if (t == null)
          this._logService.debug("[FormulaEditorShowController]: should not receive currentSkeleton$ as null!");
        else {
          const { skeleton: r } = t, s = (o = (n = this._skeleton) == null ? void 0 : n.worksheet) == null ? void 0 : o.getSheetId();
          if (this._changeRuntime(r), s !== r.worksheet.getSheetId())
            this._removeArrayFormulaRangeShape();
          else {
            const { unitId: i, sheetId: c } = t;
            this._updateArrayFormulaRangeShape(i, c);
          }
        }
      })
    );
  }
  _changeRuntime(t) {
    this._skeleton = t;
  }
  _initInterceptorEditorStart() {
    this.disposeWithMe(
      rr(
        this._sheetInterceptorService.writeCellInterceptor.intercept(os, {
          handler: (t, n, o) => {
            var a, v, p, S;
            const { row: r, col: s, unitId: i, subUnitId: c, worksheet: l } = n, g = this._formulaDataModel.getArrayFormulaRange(), m = this._formulaDataModel.getArrayFormulaCellData();
            if (this._removeArrayFormulaRangeShape(), t == null)
              return o(t);
            let h = null;
            const d = this._formulaDataModel.getFormulaStringByCell(r, s, c, i);
            if (d !== null && (h = { f: d }), t.v != null && t.v !== "" && ((p = (v = (a = m[i]) == null ? void 0 : a[c]) == null ? void 0 : v[r]) == null ? void 0 : p[s]) == null)
              return h ? { ...t, ...h } : o(t);
            const u = (S = g == null ? void 0 : g[i]) == null ? void 0 : S[c];
            return u != null && (h = this._displayArrayFormulaRangeShape(u, r, s, i, c, l, h)), h ? { ...t, ...h } : o(t);
          }
        })
      )
    );
  }
  _commandExecutedListener() {
    this.disposeWithMe(this._commandService.onCommandExecuted((t, n) => {
      (t.id === Ao.id || t.id === Fo.id && n && n.remove) && this._removeArrayFormulaRangeShape();
    })), this.disposeWithMe(
      this._commandService.beforeCommandExecuted((t) => {
        ss.id === t.id && requestIdleCallback(() => {
          const n = t.params, { unitId: o, subUnitId: r, rowsAutoHeightInfo: s } = n;
          this._refreshArrayFormulaRangeShapeByRow(o, r, s);
        });
      })
    );
  }
  _displayArrayFormulaRangeShape(t, n, o, r, s, i, c) {
    const l = this._formulaDataModel.getSheetFormulaData(r, s);
    return new Fe(t).forValue((g, m, h) => {
      var p;
      if (h == null)
        return !0;
      const { startRow: d, startColumn: u, endRow: a, endColumn: v } = h;
      if (g === n && m === o)
        return this._createArrayFormulaRangeShape(h, r), !1;
      if (n >= d && n <= a && o >= u && o <= v) {
        const S = i.getCell(d, u);
        if ((S == null ? void 0 : S.v) === ue.SPILL)
          return;
        const C = (p = l == null ? void 0 : l[g]) == null ? void 0 : p[m];
        return C == null || C.f == null ? !0 : (c == null && (c = {
          f: C.f,
          isInArrayFormulaRange: !0
        }), this._createArrayFormulaRangeShape(h, r), !1);
      }
    }), c;
  }
  _createArrayFormulaRangeShape(t, n) {
    const o = this._renderManagerService.getRenderById(n), r = this._sheetSkeletonManagerService.getCurrentSkeleton();
    if (!o || !r) return;
    const { scene: s } = o;
    if (!s) return;
    const i = {
      range: t,
      primary: null,
      style: {
        strokeWidth: 1,
        stroke: this._themeService.getColorFromTheme("primary.600"),
        fill: new un(this._themeService.getColorFromTheme("white")).setAlpha(0).toString(),
        widgets: {}
      }
    }, c = tn(i, r), { rowHeaderWidth: l, columnHeaderHeight: g } = r, m = new cr(s, Ro.FORMULA_EDITOR_SHOW, this._themeService, {
      highlightHeader: !1,
      rowHeaderWidth: l,
      columnHeaderHeight: g
    });
    m.updateRangeBySelectionWithCoord(c), m.setEvent(!1), this._previousShape = m;
  }
  _removeArrayFormulaRangeShape() {
    this._previousShape != null && (this._previousShape.dispose(), this._previousShape = null);
  }
  _refreshArrayFormulaRangeShape(t, n) {
    if (this._previousShape) {
      const { startRow: o, endRow: r, startColumn: s, endColumn: i } = this._previousShape.getRange(), c = { startRow: o, endRow: r, startColumn: s, endColumn: i };
      this._removeArrayFormulaRangeShape(), this._createArrayFormulaRangeShape(c, t);
    }
  }
  _checkCurrentSheet(t, n) {
    const o = this._sheetSkeletonManagerService.getCurrentSkeleton();
    if (!o) return !1;
    const r = o.worksheet;
    return r ? r.unitId === t && r.getSheetId() === n : !1;
  }
  _updateArrayFormulaRangeShape(t, n) {
    this._checkCurrentSheet(t, n) && this._previousShape && this._refreshArrayFormulaRangeShape(t);
  }
  _refreshArrayFormulaRangeShapeByRow(t, n, o) {
    if (!this._checkCurrentSheet(t, n) || !this._previousShape) return;
    const { startRow: r, endRow: s, startColumn: i, endColumn: c } = this._previousShape.getRange();
    for (let l = 0; l < o.length; l++) {
      const { row: g } = o[l];
      if (r >= g) {
        const m = {
          startRow: r,
          endRow: s,
          startColumn: i,
          endColumn: c
        };
        this._refreshArrayFormulaRangeShape(t, m);
        break;
      }
    }
  }
};
Vt = oi([
  He(1, q(mn)),
  He(2, q(bt)),
  He(3, q(It)),
  He(4, xe),
  He(5, q(jt)),
  He(6, me),
  He(7, eo)
], Vt);
var si = Object.getOwnPropertyDescriptor, ii = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? si(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, zn = (e, t) => (n, o) => t(n, o, e);
const ci = {
  tl: {
    size: 6,
    color: "#409f11"
  }
};
let Wt = class extends to {
  constructor(e, t) {
    super(), this._sheetInterceptorService = e, this._formulaDataModel = t, this.disposeWithMe(this._sheetInterceptorService.intercept(
      is.CELL_CONTENT,
      {
        effect: no.Style,
        handler: (n, o, r) => {
          var c, l, g, m;
          const s = (m = (g = (l = (c = this._formulaDataModel.getArrayFormulaCellData()) == null ? void 0 : c[o.unitId]) == null ? void 0 : l[o.subUnitId]) == null ? void 0 : g[o.row]) == null ? void 0 : m[o.col];
          return !lr(n, !!s) || !n || (n === o.rawData && (n = { ...o.rawData }), n.markers = {
            ...n == null ? void 0 : n.markers,
            ...ci
          }), r(n);
        },
        priority: 10
      }
    ));
  }
};
Wt = ii([
  zn(0, q(mn)),
  zn(1, q(bt))
], Wt);
function ai() {
  const e = T(us), t = T(me), n = Se(e.progress$), o = Dt(() => {
    t.executeCommand(Do.id);
  }, [t]), r = Dt(() => {
    e.clearProgress();
  }, [e]);
  return /* @__PURE__ */ M(gs, { progress: n, onTerminate: o, onClearProgress: r });
}
function li(e, t) {
  return Object.keys(e).filter((n) => isNaN(Number(n)) && n !== "DefinedName").map((n) => ({
    label: t.t(`formula.functionType.${n.toLocaleLowerCase()}`),
    value: `${e[n]}`
  }));
}
function Nr(e) {
  if (!e.require && !e.repeat)
    return `[${e.name}]`;
  if (e.require && !e.repeat)
    return e.name;
  if (!e.require && e.repeat)
    return `[${e.name},...]`;
  if (e.require && e.repeat)
    return `${e.name},...`;
}
function Or(e) {
  const { prefix: t, value: n } = e;
  return /* @__PURE__ */ W("div", { children: [
    /* @__PURE__ */ W("span", { children: [
      t,
      "("
    ] }),
    n && n.map((o, r) => /* @__PURE__ */ W("span", { children: [
      /* @__PURE__ */ M("span", { children: Nr(o) }),
      r === n.length - 1 ? "" : ","
    ] }, r)),
    ")"
  ] });
}
function vt(e) {
  const { className: t, value: n, title: o } = e;
  return /* @__PURE__ */ W("div", { className: "univer-mb-2 univer-text-xs", children: [
    /* @__PURE__ */ M(
      "div",
      {
        className: re("univer-mb-2 univer-font-medium univer-text-gray-500 dark:!univer-text-gray-300", t),
        children: o
      }
    ),
    /* @__PURE__ */ M(
      "div",
      {
        className: "univer-break-all univer-text-gray-900 dark:!univer-text-white",
        children: n
      }
    )
  ] });
}
function ui(e) {
  const { functionInfo: t, onChange: n } = e;
  if (!t) return null;
  const [o, r] = V([]), [s, i] = V(t.functionParameter), [c, l] = V(-1);
  return /* @__PURE__ */ W("div", { children: [
    /* @__PURE__ */ M("div", { className: re("univer-h-[364px] univer-overflow-y-auto", lt), children: s.map((g, m) => /* @__PURE__ */ W("div", { children: [
      /* @__PURE__ */ M("div", { className: "univer-text-sm", children: g.name }),
      /* @__PURE__ */ M("div", { className: "univer-mb-2 univer-mt-1" })
    ] }, m)) }),
    /* @__PURE__ */ M("div", { className: re("univer-flex-1 univer-p-3", Rs), children: /* @__PURE__ */ M(
      vt,
      {
        title: c === -1 ? /* @__PURE__ */ M(Or, { prefix: t.functionName, value: s }) : s[c].name,
        value: c === -1 ? t.description : s[c].detail
      }
    ) })
  ] });
}
function et({ ref: e, ...t }) {
  const { icon: n, id: o, className: r, extend: s, ...i } = t, c = `univerjs-icon univerjs-icon-${o} ${r || ""}`.trim(), l = z(`_${fi()}`);
  return xr(n, `${o}`, {
    defIds: n.defIds,
    idSuffix: l.current
  }, {
    ref: e,
    className: c,
    ...i
  }, s);
}
function xr(e, t, n, o, r) {
  return Je(e.tag, {
    key: t,
    ...di(e, n, r),
    ...o
  }, (hi(e, n).children || []).map((s, i) => xr(s, `${t}-${e.tag}-${i}`, n, void 0, r)));
}
function di(e, t, n) {
  const o = { ...e.attrs };
  n != null && n.colorChannel1 && o.fill === "colorChannel1" && (o.fill = n.colorChannel1), e.tag === "mask" && o.id && (o.id = o.id + t.idSuffix), Object.entries(o).forEach(([s, i]) => {
    s === "mask" && typeof i == "string" && (o[s] = i.replace(/url\(#(.*)\)/, `url(#$1${t.idSuffix})`));
  });
  const { defIds: r } = t;
  return !r || r.length === 0 || (e.tag === "use" && o["xlink:href"] && (o["xlink:href"] = o["xlink:href"] + t.idSuffix), Object.entries(o).forEach(([s, i]) => {
    typeof i == "string" && (o[s] = i.replace(/url\(#(.*)\)/, `url(#$1${t.idSuffix})`));
  })), o;
}
function hi(e, t) {
  var o;
  const { defIds: n } = t;
  return !n || n.length === 0 ? e : e.tag === "defs" && ((o = e.children) != null && o.length) ? {
    ...e,
    children: e.children.map((r) => typeof r.attrs.id == "string" && n && n.includes(r.attrs.id) ? {
      ...r,
      attrs: {
        ...r.attrs,
        id: r.attrs.id + t.idSuffix
      }
    } : r)
  } : e;
}
function fi() {
  return Math.random().toString(36).substring(2, 8);
}
et.displayName = "UniverIcon";
const gi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M14.1544 3.75557C14.3887 3.98988 14.3887 4.36978 14.1544 4.6041L6.51409 12.2444C6.40157 12.3569 6.24896 12.4201 6.08983 12.4201C5.9307 12.4201 5.77808 12.3569 5.66556 12.2444L1.84541 8.42425C1.6111 8.18993 1.6111 7.81003 1.84541 7.57572C2.07973 7.34141 2.45963 7.34141 2.69394 7.57572L6.08983 10.9716L13.3059 3.75557C13.5402 3.52126 13.9201 3.52126 14.1544 3.75557Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Mr = Le(function(t, n) {
  return Je(et, Object.assign({}, t, {
    id: "check-mark-icon",
    ref: n,
    icon: gi
  }));
});
Mr.displayName = "CheckMarkIcon";
const mi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M3.71274 2.86421C3.47843 2.6299 3.09853 2.6299 2.86421 2.86421C2.6299 3.09853 2.6299 3.47843 2.86421 3.71274L7.15154 8.00007L2.86421 12.2874C2.6299 12.5217 2.6299 12.9016 2.86421 13.1359C3.09853 13.3702 3.47843 13.3702 3.71274 13.1359L8.00007 8.84859L12.2874 13.1359C12.5217 13.3702 12.9016 13.3702 13.1359 13.1359C13.3702 12.9016 13.3702 12.5217 13.1359 12.2874L8.84859 8.00007L13.1359 3.71274C13.3702 3.47843 13.3702 3.09853 13.1359 2.86421C12.9016 2.6299 12.5217 2.6299 12.2874 2.86421L8.00007 7.15154L3.71274 2.86421Z"
    }
  }]
}, wr = Le(function(t, n) {
  return Je(et, Object.assign({}, t, {
    id: "close-icon",
    ref: n,
    icon: mi
  }));
});
wr.displayName = "CloseIcon";
const pi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M5.3313 1.4667C5.3313 1.13533 5.59993 0.866699 5.9313 0.866699H10.069C10.4004 0.866699 10.669 1.13533 10.669 1.4667C10.669 1.79807 10.4004 2.0667 10.069 2.0667H5.9313C5.59993 2.0667 5.3313 1.79807 5.3313 1.4667Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M1.09985 3.64443C1.09985 3.31306 1.36848 3.04443 1.69985 3.04443H14.2999C14.6312 3.04443 14.8999 3.31306 14.8999 3.64443C14.8999 3.9758 14.6312 4.24443 14.2999 4.24443H1.69985C1.36848 4.24443 1.09985 3.9758 1.09985 3.64443Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M6.12398 8.30171C6.35829 8.0674 6.73819 8.0674 6.97251 8.30171L8.00007 9.32928L9.02764 8.30171C9.26195 8.0674 9.64185 8.0674 9.87617 8.30171C10.1105 8.53603 10.1105 8.91593 9.87617 9.15024L8.8486 10.1778L9.87617 11.2054C10.1105 11.4397 10.1105 11.8196 9.87617 12.0539C9.64185 12.2882 9.26195 12.2882 9.02764 12.0539L8.00007 11.0263L6.97251 12.0539C6.73819 12.2882 6.35829 12.2882 6.12398 12.0539C5.88966 11.8196 5.88966 11.4397 6.12398 11.2054L7.15154 10.1778L6.12398 9.15024C5.88966 8.91593 5.88966 8.53603 6.12398 8.30171Z"
      }
    },
    {
      tag: "path",
      attrs: {
        fill: "currentColor",
        d: "M4.75332 5.22217C3.86966 5.22217 3.15332 5.93851 3.15332 6.82217V12.5331C3.15332 13.9691 4.31738 15.1332 5.75332 15.1332H10.2465C11.6825 15.1332 12.8465 13.9691 12.8465 12.5331V6.82217C12.8465 5.93851 12.1302 5.22217 11.2465 5.22217H4.75332ZM4.35332 6.82217C4.35332 6.60125 4.53241 6.42217 4.75332 6.42217H11.2465C11.4674 6.42217 11.6465 6.60125 11.6465 6.82217V12.5331C11.6465 13.3063 11.0197 13.9332 10.2465 13.9332H5.75332C4.98012 13.9332 4.35332 13.3063 4.35332 12.5331V6.82217Z",
        fillRule: "evenodd",
        clipRule: "evenodd"
      }
    }
  ]
}, kr = Le(function(t, n) {
  return Je(et, Object.assign({}, t, {
    id: "delete-icon",
    ref: n,
    icon: pi
  }));
});
kr.displayName = "DeleteIcon";
const Si = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M8.6 1.99991C8.60001 1.66854 8.33138 1.39991 8.00001 1.3999C7.66864 1.3999 7.40001 1.66853 7.4 1.9999L7.39996 7.3999H1.9999C1.66853 7.3999 1.3999 7.66853 1.3999 7.9999C1.3999 8.33127 1.66853 8.5999 1.9999 8.5999H7.39995L7.3999 13.9999C7.3999 14.3313 7.66853 14.5999 7.9999 14.5999C8.33127 14.5999 8.5999 14.3313 8.5999 13.9999L8.59995 8.5999H13.9999C14.3313 8.5999 14.5999 8.33127 14.5999 7.9999C14.5999 7.66853 14.3313 7.3999 13.9999 7.3999H8.59996L8.6 1.99991Z"
    }
  }]
}, Ar = Le(function(t, n) {
  return Je(et, Object.assign({}, t, {
    id: "increase-icon",
    ref: n,
    icon: Si
  }));
});
Ar.displayName = "IncreaseIcon";
const vi = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M5.90913 3.57564C6.14345 3.34132 6.52335 3.34132 6.75766 3.57564L10.7577 7.57564C10.992 7.80995 10.992 8.18985 10.7577 8.42417L6.75766 12.4242C6.52335 12.6585 6.14345 12.6585 5.90913 12.4242C5.67482 12.1899 5.67482 11.81 5.90913 11.5756L9.48487 7.9999L5.90913 4.42417C5.67482 4.18985 5.67482 3.80995 5.90913 3.57564Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Fr = Le(function(t, n) {
  return Je(et, Object.assign({}, t, {
    id: "more-icon",
    ref: n,
    icon: vi
  }));
});
Fr.displayName = "MoreIcon";
const Ci = {
  tag: "svg",
  attrs: {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    viewBox: "0 0 16 16",
    width: "1em",
    height: "1em"
  },
  children: [{
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M12.6185 12.4423C12.5907 12.2749 12.7773 12.15 12.9343 12.2308L15.4242 13.5126C15.6102 13.6084 15.5544 13.8745 15.3439 13.8955L14.2456 14.184L13.4521 15.1286C13.3495 15.2939 13.085 15.2463 13.0534 15.0568L12.6185 12.4423Z"
    }
  }, {
    tag: "path",
    attrs: {
      fill: "currentColor",
      d: "M1 3.6C1 2.16406 2.16406 1 3.6 1H12.3C13.7359 1 14.9 2.16406 14.9 3.6V5.81156C14.9003 5.81881 14.9004 5.82609 14.9004 5.8334C14.9004 5.84071 14.9003 5.84799 14.9 5.85524V10.045C14.9003 10.0522 14.9004 10.0595 14.9004 10.0668C14.9004 10.3982 14.6318 10.6668 14.3004 10.6668H11.1668C10.8907 10.6668 10.6668 10.8907 10.6668 11.1668V14.3C10.6668 14.6314 10.3982 14.9 10.0668 14.9L10.05 14.8998L3.6 14.9C2.16406 14.9 1 13.7359 1 12.3V3.6ZM13.2 5.2334C13.4761 5.2334 13.7 5.00954 13.7 4.7334V3.6C13.7 2.8268 13.0732 2.2 12.3 2.2H11.1668C10.8907 2.2 10.6668 2.42386 10.6668 2.7V4.7334C10.6668 5.00954 10.8907 5.2334 11.1668 5.2334H13.2ZM10.6668 6.9334C10.6668 6.65726 10.8907 6.4334 11.1668 6.4334H13.2C13.4761 6.4334 13.7 6.65726 13.7 6.9334V8.9668C13.7 9.24294 13.4761 9.4668 13.2 9.4668H11.1668C10.8907 9.4668 10.6668 9.24294 10.6668 8.9668V6.9334ZM8.9668 5.2334C9.24294 5.2334 9.4668 5.00954 9.4668 4.7334V2.7C9.4668 2.42386 9.24294 2.2 8.9668 2.2H6.9334C6.65726 2.2 6.4334 2.42386 6.4334 2.7V4.7334C6.4334 5.00954 6.65726 5.2334 6.9334 5.2334L8.9668 5.2334ZM6.4334 6.9334C6.4334 6.65726 6.65726 6.4334 6.9334 6.4334L8.9668 6.4334C9.24294 6.4334 9.4668 6.65726 9.4668 6.9334V8.9668C9.4668 9.24294 9.24294 9.4668 8.9668 9.4668L6.9334 9.4668C6.65726 9.4668 6.4334 9.24294 6.4334 8.9668V6.9334ZM4.7334 5.2334C5.00954 5.2334 5.2334 5.00954 5.2334 4.7334V2.7C5.2334 2.42386 5.00954 2.2 4.7334 2.2H3.6C2.8268 2.2 2.2 2.8268 2.2 3.6V4.7334C2.2 5.00954 2.42386 5.2334 2.7 5.2334H4.7334ZM2.2 6.9334C2.2 6.65726 2.42386 6.4334 2.7 6.4334H4.7334C5.00954 6.4334 5.2334 6.65725 5.2334 6.9334V8.9668C5.2334 9.24294 5.00954 9.4668 4.7334 9.4668H2.7C2.42386 9.4668 2.2 9.24294 2.2 8.9668V6.9334ZM5.2334 11.1668C5.2334 10.8907 5.00954 10.6668 4.7334 10.6668H2.7C2.42386 10.6668 2.2 10.8907 2.2 11.1668V12.3C2.2 13.0732 2.8268 13.7 3.6 13.7H4.7334C5.00954 13.7 5.2334 13.4761 5.2334 13.2V11.1668ZM9.4668 11.1668C9.4668 10.8907 9.24294 10.6668 8.9668 10.6668H6.9334C6.65726 10.6668 6.4334 10.8907 6.4334 11.1668V13.2C6.4334 13.4761 6.65726 13.7 6.9334 13.7H8.9668C9.24294 13.7 9.4668 13.4761 9.4668 13.2V11.1668Z",
      fillRule: "evenodd",
      clipRule: "evenodd"
    }
  }]
}, Dr = Le(function(t, n) {
  return Je(et, Object.assign({}, t, {
    id: "select-range-icon",
    ref: n,
    icon: Ci
  }));
});
Dr.displayName = "SelectRangeIcon";
function _i(e) {
  const { onChange: t } = e, n = "-1", [o, r] = V(""), [s, i] = V([]), [c, l] = V(0), [g, m] = V(n), [h, d] = V(0), [u, a] = V(null), v = T(Yt), p = T(dt), S = T(pn), C = Se(S.sidebarOptions$), _ = li(ur, p);
  _.unshift({
    label: p.t("formula.moreFunctions.allFunctions"),
    value: n
  });
  const E = p.t("formula.prompt.required"), f = p.t("formula.prompt.optional");
  $(() => {
    A(n);
  }, []), $(() => {
    y(0);
  }, [s]), $(() => {
    C != null && C.visible && (r(""), i([]), l(0), m(n), d(0), a(null), A(n));
  }, [C]);
  const b = (I) => {
    if (o.trim() === "") return I;
    const x = new RegExp(`(${o.toLocaleUpperCase()})`);
    return I.split(x).filter(Boolean).map((w, D) => w.match(x) ? /* @__PURE__ */ M("span", { className: "univer-text-red-500", children: w }, D) : w);
  }, y = (I) => {
    if (s.length === 0) {
      a(null);
      return;
    }
    d(I);
    const x = v.getFunctionInfo(s[I].name);
    if (!x) {
      a(null);
      return;
    }
    a(x), t(x);
  };
  function A(I) {
    m(I);
    const x = v.getSearchListByType(+I);
    i(x);
  }
  function F(I) {
    r(I);
    const x = v.getSearchListByName(I);
    i(x);
  }
  function N(I) {
    if (I.stopPropagation(), I.key === "ArrowDown") {
      const x = c + 1;
      l(x === s.length ? 0 : x);
    } else if (I.key === "ArrowUp") {
      const x = c - 1;
      l(x === -1 ? s.length - 1 : x);
    } else I.key === "Enter" && y(c);
  }
  const O = (I) => {
    l(I);
  }, k = () => {
    l(-1);
  };
  return /* @__PURE__ */ W("div", { children: [
    /* @__PURE__ */ W("div", { className: "univer-flex univer-items-center univer-justify-between univer-gap-2", children: [
      /* @__PURE__ */ M(Is, { value: g, options: _, onChange: A }),
      /* @__PURE__ */ M(
        _r,
        {
          placeholder: p.t("formula.moreFunctions.searchFunctionPlaceholder"),
          onKeyDown: N,
          value: o,
          onChange: F,
          size: "small",
          allowClear: !0
        }
      )
    ] }),
    s.length > 0 && /* @__PURE__ */ M(
      "ul",
      {
        className: re("univer-mb-0 univer-mt-2 univer-box-border univer-max-h-72 univer-w-full univer-select-none univer-list-none univer-overflow-y-auto univer-rounded univer-p-3 univer-outline-none", Sn, lt),
        onKeyDown: N,
        tabIndex: -1,
        children: s.map(({ name: I }, x) => /* @__PURE__ */ W(
          "li",
          {
            className: re("univer-relative univer-box-border univer-cursor-pointer univer-rounded univer-px-7 univer-py-1 univer-text-sm univer-text-gray-900 univer-transition-colors dark:!univer-text-white", {
              "univer-bg-gray-200 dark:!univer-bg-gray-600": c === x
            }),
            onMouseEnter: () => O(x),
            onMouseLeave: k,
            onClick: () => y(x),
            children: [
              h === x && /* @__PURE__ */ M(
                Mr,
                {
                  className: "univer-absolute univer-left-1.5 univer-top-1/2 univer-inline-flex -univer-translate-y-1/2 univer-text-base univer-text-primary-600"
                }
              ),
              /* @__PURE__ */ M("span", { className: "univer-block", children: b(I) })
            ]
          },
          x
        ))
      }
    ),
    u && /* @__PURE__ */ W("div", { className: re("univer-mx-0 univer-my-2 univer-overflow-y-auto", lt), children: [
      /* @__PURE__ */ M(vt, { title: u.functionName, value: u.description }),
      /* @__PURE__ */ M(
        vt,
        {
          title: p.t("formula.moreFunctions.syntax"),
          value: /* @__PURE__ */ M(Or, { prefix: u.functionName, value: u.functionParameter })
        }
      ),
      /* @__PURE__ */ M(
        vt,
        {
          title: p.t("formula.prompt.helpExample"),
          value: `${u.functionName}(${u.functionParameter.map((I) => I.example).join(",")})`
        }
      ),
      u.functionParameter && u.functionParameter.map((I) => /* @__PURE__ */ M(
        vt,
        {
          title: I.name,
          value: `${I.require ? E : f} ${I.detail}`
        },
        I.name
      ))
    ] })
  ] });
}
function Ri() {
  const e = Io(), [t, n] = V(!0), [o, r] = V(!1), [s, i] = V(null);
  T(dn);
  const c = T(dt), l = T(Ye), g = T(ne), m = T(me);
  function h() {
    n(!t), r(!o);
  }
  function d() {
    const u = pr(g);
    if (!u) return;
    m.executeCommand(sr.id, {
      visible: !0,
      unitId: u.unitId,
      eventType: Oe.Dblclick
    });
    const a = l.getEditor(an), v = l.getEditor(tr), p = `=${s == null ? void 0 : s.functionName}(`;
    a == null || a.replaceText(p), v == null || v.replaceText(p, !1);
  }
  return /* @__PURE__ */ W(
    "div",
    {
      "data-u-comp": "sheets-formula-functions-panel",
      className: "univer-box-border univer-flex univer-h-full univer-flex-col univer-justify-between univer-py-2",
      children: [
        t && /* @__PURE__ */ M(_i, { onChange: i }),
        o && /* @__PURE__ */ M(ui, { functionInfo: s, onChange: () => {
        } }),
        /* @__PURE__ */ W("div", { className: "univer-flex univer-justify-end", children: [
          o && /* @__PURE__ */ M(
            ct,
            {
              variant: "primary",
              onClick: h,
              className: "univer-mb-5 univer-ml-4 univer-mr-0 univer-mt-0",
              children: c.t("formula.moreFunctions.next")
            }
          ),
          o && /* @__PURE__ */ M(ct, { onClick: h, className: "univer-mb-5 univer-ml-4 univer-mr-0 univer-mt-0", children: c.t("formula.moreFunctions.prev") }),
          t && !!e && /* @__PURE__ */ M(
            ct,
            {
              variant: "primary",
              onClick: d,
              className: "univer-mb-5 univer-ml-4 univer-mr-0 univer-mt-0",
              children: c.t("formula.moreFunctions.confirm")
            }
          )
        ] })
      ]
    }
  );
}
function Ii(e) {
  return {
    id: pe.id,
    title: "SUM",
    icon: "SumIcon",
    type: Qe.BUTTON,
    params: {
      value: "SUM"
    },
    hidden$: ft(e, H.UNIVER_SHEET),
    disabled$: Ke(e, {
      workbookTypes: [Xe],
      worksheetTypes: [Ze, Ge],
      rangeTypes: [ze]
    })
  };
}
function Ei(e) {
  return {
    id: pe.id,
    title: "COUNT",
    icon: "CntIcon",
    type: Qe.BUTTON,
    params: {
      value: "COUNT"
    },
    hidden$: ft(e, H.UNIVER_SHEET),
    disabled$: Ke(e, {
      workbookTypes: [Xe],
      worksheetTypes: [Ze, Ge],
      rangeTypes: [ze]
    })
  };
}
function bi(e) {
  return {
    id: pe.id,
    title: "AVERAGE",
    icon: "AvgIcon",
    type: Qe.BUTTON,
    params: {
      value: "AVERAGE"
    },
    hidden$: ft(e, H.UNIVER_SHEET),
    disabled$: Ke(e, {
      workbookTypes: [Xe],
      worksheetTypes: [Ze, Ge],
      rangeTypes: [ze]
    })
  };
}
function yi(e) {
  return {
    id: pe.id,
    title: "MAX",
    icon: "MaxIcon",
    type: Qe.BUTTON,
    params: {
      value: "MAX"
    },
    hidden$: ft(e, H.UNIVER_SHEET),
    disabled$: Ke(e, {
      workbookTypes: [Xe],
      worksheetTypes: [Ze, Ge],
      rangeTypes: [ze]
    })
  };
}
function Ti(e) {
  return {
    id: pe.id,
    title: "MIN",
    icon: "MinIcon",
    type: Qe.BUTTON,
    params: {
      value: "MIN"
    },
    hidden$: ft(e, H.UNIVER_SHEET),
    disabled$: Ke(e, {
      workbookTypes: [Xe],
      worksheetTypes: [Ze, Ge],
      rangeTypes: [ze]
    })
  };
}
function Ni(e) {
  return {
    id: Cn.id,
    title: "formula.insert.more",
    tooltip: "formula.insert.tooltip",
    type: Qe.BUTTON,
    hidden$: ft(e, H.UNIVER_SHEET),
    disabled$: Ke(e, {
      workbookTypes: [Xe],
      worksheetTypes: [Ze, Ge],
      rangeTypes: [ze]
    })
  };
}
function Oi(e) {
  return e.get(ne).getCurrentTypeOfUnit$(H.UNIVER_SHEET).pipe(
    qo((o) => o ? e.get(qn) ? new jo((s) => s.next(!e.get(qn).supportClipboard)) : Pn(!0) : Pn(!0))
  );
}
function xi(e) {
  return {
    id: vn.id,
    type: Qe.BUTTON,
    title: "formula.operation.pasteFormula",
    disabled$: Oi(e).pipe(
      Bo(Ke(e, {
        workbookTypes: [Xe],
        rangeTypes: [ze],
        worksheetTypes: [Ge, Ze]
      })),
      hr(([t, n]) => t || n)
    )
  };
}
const Mi = {
  [jn.BASIC]: {
    [`${pe.id}.sum`]: {
      order: 0,
      menuItemFactory: Ii
    },
    [`${pe.id}.count`]: {
      order: 1,
      menuItemFactory: Ei
    },
    [`${pe.id}.average`]: {
      order: 2,
      menuItemFactory: bi
    },
    [`${pe.id}.max`]: {
      order: 3,
      menuItemFactory: yi
    },
    [`${pe.id}.min`]: {
      order: 4,
      menuItemFactory: Ti
    }
  },
  [jn.OTHERS]: {
    [Cn.id]: {
      order: 0,
      menuItemFactory: Ni
    }
  },
  [Eo]: {
    [vn.id]: {
      order: 4,
      menuItemFactory: xi
    }
  }
}, wi = "meta_key_ctrl_And_Shift";
function ki(e) {
  return e.getContextValue(ro) && e.getContextValue(oo);
}
const zt = [
  L.ARROW_DOWN,
  L.ARROW_UP,
  L.ARROW_LEFT,
  L.ARROW_RIGHT
], Ai = [...zt, L.ENTER, L.TAB, L.ESC];
function Fi() {
  const e = [];
  for (const t of Ai)
    e.push({
      id: gt.id,
      binding: t,
      preconditions: (n) => Et(n),
      staticParameters: {
        eventType: Oe.Keyboard,
        keycode: t
      }
    });
  return e;
}
function Di() {
  const e = [];
  for (const t of zt)
    e.push({
      id: gt.id,
      binding: t | P.SHIFT,
      preconditions: (n) => Et(n),
      staticParameters: {
        eventType: Oe.Keyboard,
        keycode: t,
        metaKey: P.SHIFT
      }
    });
  return e;
}
function Li() {
  const e = [];
  for (const t of zt)
    e.push({
      id: gt.id,
      binding: t | P.CTRL_COMMAND,
      preconditions: (n) => Et(n),
      staticParameters: {
        eventType: Oe.Keyboard,
        keycode: t,
        metaKey: P.CTRL_COMMAND
      }
    });
  return e;
}
function $i() {
  const e = [];
  for (const t of zt)
    e.push({
      id: gt.id,
      binding: t | P.SHIFT | P.CTRL_COMMAND,
      preconditions: (n) => Et(n),
      staticParameters: {
        eventType: Oe.Keyboard,
        keycode: t,
        metaKey: wi
      }
    });
  return e;
}
const Pi = {
  id: br.id,
  binding: L.F4,
  preconditions: (e) => Et(e)
};
function Ui() {
  const e = [];
  for (const t of [L.ENTER, L.TAB, L.ARROW_DOWN, L.ARROW_UP])
    e.push({
      id: gt.id,
      binding: t,
      preconditions: (n) => ki(n),
      staticParameters: {
        eventType: Oe.Keyboard,
        keycode: t,
        isSingleEditor: !0
      }
    });
  return e;
}
const Vi = {
  id: ds.id,
  binding: P.ALT | L.EQUAL,
  preconditions: bo,
  mac: P.CTRL_COMMAND | P.ALT | L.EQUAL,
  description: "shortcut.sheets-formula-ui.quick-sum",
  group: "4_sheet-edit"
};
var Wi = Object.getOwnPropertyDescriptor, Hi = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Wi(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, Be = (e, t) => (n, o) => t(n, o, e);
let Ht = class extends ut {
  constructor(e, t, n, o, r, s, i) {
    super(), this._injector = e, this._menuManagerService = t, this._commandService = n, this._shortcutService = o, this._uiPartsService = r, this._renderManagerService = s, this._componentManager = i, this._initialize();
  }
  _initialize() {
    this._registerCommands(), this._registerMenus(), this._registerShortcuts(), this._registerComponents(), this._registerRenderModules();
  }
  _registerMenus() {
    this._menuManagerService.mergeMenu(Mi);
  }
  _registerCommands() {
    [
      vn,
      pe,
      Cn,
      Ps,
      ws,
      gt,
      br
    ].forEach((e) => this.disposeWithMe(this._commandService.registerCommand(e)));
  }
  _registerShortcuts() {
    [
      ...Fi(),
      ...Di(),
      ...Li(),
      ...$i(),
      ...Ui(),
      Vi,
      Pi
    ].forEach((e) => {
      this.disposeWithMe(this._shortcutService.registerShortcut(e));
    });
  }
  _registerComponents() {
    this.disposeWithMe(this._uiPartsService.registerComponent(yo.FORMULA_AUX, () => Cr(ai, this._injector))), this._componentManager.register(Er, Ri);
  }
  _registerRenderModules() {
    this.disposeWithMe(this._renderManagerService.registerRenderModule(H.UNIVER_SHEET, [Vt]));
  }
};
Ht = Hi([
  Be(0, q(Rt)),
  Be(1, ms),
  Be(2, me),
  Be(3, Gt),
  Be(4, Sr),
  Be(5, xe),
  Be(6, q(vr))
], Ht);
class Lr {
  constructor() {
    j(this, "_currentSelector$", new fr(null));
    j(this, "currentSelector$", this._currentSelector$.asObservable());
  }
  showRangeSelectorDialog(t) {
    const n = t.callback, o = new Promise((r) => {
      t.callback = (s) => {
        r(s), n(s);
      };
    });
    return this._currentSelector$.next(t), o;
  }
}
var Bi = Object.getOwnPropertyDescriptor, qi = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Bi(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, st = (e, t) => (n, o) => t(n, o, e);
let De = class extends To {
  constructor(t, n, o, r, s, i, c) {
    super(
      n,
      o,
      r,
      s,
      i
    );
    j(this, "_workbookSelections");
    j(this, "_eventDisposables");
    this._context = t, this._contextService = i, this._refSelectionsService = c, this._workbookSelections = this._refSelectionsService.getWorkbookSelections(this._context.unitId), this._initSelectionChangeListener(), this._initSkeletonChangeListener(), this._initUserActionSyncListener(), this._setSelectionStyle(ji(this._themeService)), this._remainLastEnabled = !0, this._highlightHeader = !1;
  }
  getLocation() {
    return this._skeleton.getLocation();
  }
  setRemainLastEnabled(t) {
    this._remainLastEnabled = t;
  }
  /**
   * This is set to true when you need to add a new selection.
   * @param {boolean} enabled
   * @memberof RefSelectionsRenderService
   */
  setSkipLastEnabled(t) {
    this._skipLastEnabled = t;
  }
  clearLastSelection() {
    const t = this._selectionControls[this._selectionControls.length - 1];
    t && (t.dispose(), this._selectionControls.pop());
  }
  /**
   * Call this method and user will be able to select on the canvas to update selections.
   */
  enableSelectionChanging() {
    return this._disableSelectionChanging(), this._eventDisposables = this._initCanvasEventListeners(), rr(() => this._disableSelectionChanging());
  }
  _disableSelectionChanging() {
    var t;
    (t = this._eventDisposables) == null || t.dispose(), this._eventDisposables = null;
  }
  disableSelectionChanging() {
    this._disableSelectionChanging();
  }
  _initCanvasEventListeners() {
    const t = this._getSheetObject(), { spreadsheetRowHeader: n, spreadsheetColumnHeader: o, spreadsheet: r, spreadsheetLeftTopPlaceholder: s } = t, { scene: i } = this._context, c = new qe();
    return c.add(r == null ? void 0 : r.onPointerDown$.subscribeEvent((l, g) => {
      this.inRefSelectionMode() && (this._onPointerDown(l, r.zIndex + 1, ve.NORMAL, this._getActiveViewport(l)), l.button !== 2 && g.stopPropagation());
    })), c.add(
      n == null ? void 0 : n.onPointerDown$.subscribeEvent((l, g) => {
        if (!this.inRefSelectionMode()) return;
        const m = this._sheetSkeletonManagerService.getCurrent().skeleton, { row: h } = An(l.offsetX, l.offsetY, i, m);
        Fn(this._workbookSelections.getCurrentSelections(), h, ve.ROW) || (this._onPointerDown(l, (r.zIndex || 1) + 1, ve.ROW, this._getActiveViewport(l), Xt.Y), l.button !== 2 && g.stopPropagation());
      })
    ), c.add(o == null ? void 0 : o.onPointerDown$.subscribeEvent((l, g) => {
      if (!this.inRefSelectionMode()) return;
      const m = this._sheetSkeletonManagerService.getCurrent().skeleton, { column: h } = An(l.offsetX, l.offsetY, i, m);
      Fn(this._workbookSelections.getCurrentSelections(), h, ve.COLUMN) || (this._onPointerDown(l, (r.zIndex || 1) + 1, ve.COLUMN, this._getActiveViewport(l), Xt.X), l.button !== 2 && g.stopPropagation());
    })), c.add(s == null ? void 0 : s.onPointerDown$.subscribeEvent((l, g) => {
      if (this._reset(), !this.inRefSelectionMode()) return;
      const m = this._sheetSkeletonManagerService.getCurrent().skeleton, h = No(m);
      this._addSelectionControlByModelData(h), this._selectionMoveStart$.next(this.getSelectionDataWithStyle());
      const d = i.onPointerUp$.subscribeEvent(() => {
        d.unsubscribe(), this._selectionMoveEnd$.next(this.getSelectionDataWithStyle());
      });
      l.button !== 2 && g.stopPropagation();
    })), c;
  }
  /**
   * Add a selection in spreadsheet, create a new SelectionControl and then update this control by range derives from selection.
   * For ref selection, create selectionShapeExtension to handle user action.
   * @param {ISelectionWithCoord} selectionWithStyle
   */
  _addSelectionControlByModelData(t) {
    var i;
    const n = this._skeleton, o = (i = t.style) != null ? i : ar(this._themeService), r = this._scene;
    return t.style = o, this.newSelectionControl(r, n, t);
  }
  _initSelectionChangeListener() {
    this.disposeWithMe(this._refSelectionsService.selectionSet$.subscribe((t) => {
      this._reset(), this._skeleton && this.resetSelectionsByModelData(t || []);
    }));
  }
  /**
   * Update selectionModel in this._workbookSelections by user action in spreadsheet area.
   */
  _initUserActionSyncListener() {
    this.disposeWithMe(this.selectionMoveStart$.subscribe((t) => {
      this._updateSelections(t, Qt.MOVE_START);
    })), this.disposeWithMe(this.selectionMoving$.subscribe((t) => {
      this._updateSelections(t, Qt.MOVING);
    })), this.disposeWithMe(this.selectionMoveEnd$.subscribe((t) => {
      this._updateSelections(t, Qt.MOVE_END);
    }));
  }
  _updateSelections(t, n) {
    const r = this._context.unit.getActiveSheet().getSheetId();
    t.length !== 0 && this._workbookSelections.setSelections(
      r,
      t.map((s) => cs(s)),
      n
    );
  }
  _initSkeletonChangeListener() {
    this.disposeWithMe(this._sheetSkeletonManagerService.currentSkeleton$.subscribe((t) => {
      var i;
      if (!t)
        return;
      const { skeleton: n } = t, { scene: o } = this._context, r = o.getViewport(Vn.VIEW_MAIN);
      this._skeleton && ((i = this._skeleton.worksheet) == null ? void 0 : i.getSheetId()) !== n.worksheet.getSheetId() && this._reset(), this._changeRuntime(n, o, r);
      const s = this._workbookSelections.getCurrentSelections();
      this.resetSelectionsByModelData(s);
    }));
  }
  _getActiveViewport(t) {
    const n = this._getSheetObject();
    return n == null ? void 0 : n.scene.getActiveViewportByCoord(Wn.FromArray([t.offsetX, t.offsetY]));
  }
  _getSheetObject() {
    return Oo(this._context.unit, this._context);
  }
  /**
   * Handle pointer down event, bind pointermove & pointerup handler.
   * then trigger selectionMoveStart$.
   *
   * @param evt
   * @param _zIndex
   * @param rangeType
   * @param viewport
   * @param scrollTimerType
   */
  // eslint-disable-next-line complexity, max-lines-per-function
  _onPointerDown(t, n = 0, o = ve.NORMAL, r, s = Xt.ALL) {
    var N;
    this._rangeType = o;
    const i = this._skeleton, c = this._scene;
    if (!c || !i)
      return;
    r && (this._activeViewport = r);
    const { offsetX: l, offsetY: g } = t, m = c.getViewport(Vn.VIEW_MAIN);
    if (!m) return;
    const h = c.getCoordRelativeToViewport(Wn.FromArray([l, g])), { x: d, y: u } = h;
    this._startViewportPosX = d, this._startViewportPosY = u;
    const a = c.getScrollXYInfoByViewport(h), { scaleX: v, scaleY: p } = c.getAncestorScale(), S = this._skeleton.getCellByOffset(d, u, v, p, a);
    if (!S) return;
    switch (o) {
      case ve.NORMAL:
        break;
      case ve.ROW:
        S.startColumn = 0, S.endColumn = this._skeleton.getColumnCount() - 1;
        break;
      case ve.COLUMN:
        S.startRow = 0, S.endRow = this._skeleton.getRowCount() - 1;
        break;
      case ve.ALL:
        S.startRow = 0, S.startColumn = 0, S.endRow = this._skeleton.getRowCount() - 1, S.endColumn = this._skeleton.getColumnCount() - 1;
    }
    const C = { range: S, primary: S, style: null };
    C.range.rangeType = o;
    const _ = tn(C, this._skeleton);
    this._startRangeWhenPointerDown = { ..._.rangeWithCoord };
    const E = { ..._.rangeWithCoord, rangeType: o };
    let f = this.getActiveSelectionControl();
    const b = this.getSelectionControls();
    for (const O of b) {
      if (t.button === 2 && er.contains(O.model, E)) {
        f = O;
        return;
      }
      if (O.model.isEqual(E)) {
        f = O;
        break;
      }
    }
    this._checkClearPreviousControls(t);
    const y = f == null ? void 0 : f.model.currentCell, A = t.shiftKey && y, F = this._remainLastEnabled && !t.ctrlKey && !t.shiftKey && !this._skipLastEnabled && !this._singleSelectionEnabled;
    A && y ? this._makeSelectionByTwoCells(
      y,
      E,
      i,
      o,
      f
      // Get updated in this method
    ) : F && f ? f.updateRangeBySelectionWithCoord(_) : f = this.newSelectionControl(c, i, C);
    for (let O = 0; O < this.getSelectionControls().length - 1; O++)
      this.getSelectionControls()[O].clearHighlight();
    this._selectionMoveStart$.next(this.getSelectionDataWithStyle()), c.disableObjectsEvent(), this._clearUpdatingListeners(), this._addEndingListeners(), (N = c.getTransformer()) == null || N.clearSelectedObjects(), this._setupPointerMoveListener(m, f, o, s, d, u), this._escapeShortcutDisposable = this._shortcutService.forceEscape(), this._scenePointerUpSub = c.onPointerUp$.subscribeEvent(() => {
      var O;
      this._clearUpdatingListeners(), this._selectionMoveEnd$.next(this.getSelectionDataWithStyle()), (O = this._escapeShortcutDisposable) == null || O.dispose(), this._escapeShortcutDisposable = null;
    });
  }
  /**
   * Diff between normal selection, no highlightHeader for ref selections.
   * @param scene
   * @param skeleton
   * @param selectionWithCoord
   * @returns {SelectionControl} selectionControl just created
   */
  newSelectionControl(t, n, o) {
    const r = this.getSelectionControls().length, { rowHeaderWidth: s, columnHeaderHeight: i } = n, c = new cr(t, r, this._themeService, {
      highlightHeader: this._highlightHeader,
      enableAutoFill: !1,
      rowHeaderWidth: s,
      columnHeaderHeight: i
    }), l = tn(o, n);
    return c.updateRangeBySelectionWithCoord(l), this._selectionControls.push(c), c.setControlExtension({
      skeleton: n,
      scene: t,
      themeService: this._themeService,
      injector: this._injector,
      selectionHooks: {
        selectionMoveEnd: () => {
          this._selectionMoveEnd$.next(this.getSelectionDataWithStyle());
        }
      }
    }), c;
  }
};
De = qi([
  st(1, q(Rt)),
  st(2, q(It)),
  st(3, Gt),
  st(4, q(jt)),
  st(5, cn),
  st(6, Kt)
], De);
function ji(e) {
  const t = ar(e);
  return t.widgets = { tl: !0, tc: !0, tr: !0, ml: !0, mr: !0, bl: !0, bc: !0, br: !0 }, t;
}
const _n = (e, t, n = !0) => {
  let o = -1;
  return e.reduce((r, s, i) => {
    if (r.isFinish)
      return r;
    const c = r.currentIndex;
    if (typeof s != "string")
      r.currentIndex += s.token.length;
    else {
      const l = s.length;
      r.currentIndex += l;
    }
    return (n ? r.currentIndex === t : t > c && t <= r.currentIndex) && (o = i, r.isFinish = !0), r;
  }, { currentIndex: 0, isFinish: !1 }), o;
}, $r = (e, t) => {
  const n = e[t];
  let o = -1;
  if (!n || typeof n == "string" || n.nodeType !== Q.REFERENCE) return -1;
  for (let r = 0; r <= t; r++) {
    const s = e[r];
    typeof s != "string" && s.nodeType === Q.REFERENCE && o++;
  }
  return o;
}, Ki = (e, t = 100) => {
  $(() => {
    let n = null;
    const o = () => {
      n === null && (n = window.setTimeout(() => {
        e(), n = null;
      }, t));
    };
    return window.addEventListener("scroll", o), window.addEventListener("resize", o), () => {
      n !== null && clearTimeout(n), window.removeEventListener("scroll", o), window.removeEventListener("resize", o);
    };
  }, [e, t]);
};
function Pr(e, t, n) {
  const o = T(Ye), r = ie(() => new fr({ left: -999, top: -999, right: -999, bottom: -999 }), []), s = T(pn), i = T(ne), c = te(() => {
    var _;
    const l = o.getEditor(e);
    if (!l)
      return;
    const g = l.getBoundingClientRect(), { marginTop: m = 0, marginBottom: h = 0 } = l.getDocumentData().documentStyle, d = l.getSkeleton();
    if (!d) return;
    const u = (_ = d.getSkeletonData()) == null ? void 0 : _.pages[0].height;
    let { left: a, top: v, right: p, bottom: S } = g;
    v = v + m, S = u ? v + u : S - h;
    const C = r.getValue();
    if (!(C.left === a && C.top === v && C.right === p && C.bottom === S))
      return r.next({ left: a - 1, right: p + 1, top: v - 1, bottom: S + 1 }), g;
  });
  return $(() => {
    t && c();
  }, [e, o, i.unitAdded$, c, t, ...n != null ? n : []]), Ki(c), $(() => {
    const l = s.scrollEvent$.pipe(Ko(100)).subscribe(c);
    return () => {
      l.unsubscribe();
    };
  }, []), [r, c];
}
const tt = (e) => {
  const t = z(e);
  return t.current = e, t;
}, Yi = (e, t, n) => {
  const o = T(Zt), r = T(Yt), s = T(_e), [i, c] = V(), [l, g] = V(-1), [m, h] = V(!0), d = tt(m), u = z(t);
  u.current = t;
  const a = () => {
    c(void 0), g(-1), h(!1);
  };
  return $(() => {
    const v = s.sequenceNodesBuilder(t.slice(1));
    o.setSequenceNodes(v != null ? v : []);
  }, [t]), $(() => {
    if (n && e) {
      const v = n.selectionChange$.pipe(hn(50)).subscribe((S) => {
        if (S.textRanges.length === 1) {
          const [C] = S.textRanges;
          if (C.collapsed && d.current) {
            const { startOffset: _ } = C, E = o.getCurrentSequenceNodeIndex(_ - 2), f = o.getCurrentSequenceNodeByIndex(E), b = o.getCurrentSequenceNodeByIndex(E + 1);
            if (f)
              if (typeof f != "string" && f.nodeType === 3 && !r.hasDefinedNameDescription(f.token.trim()) && b === je.OPEN_BRACKET) {
                const y = r.getFunctionInfo(f.token);
                c(y), g(-1);
                return;
              } else {
                const y = s.getFunctionAndParameter(`${u.current}A`, _ - 1);
                if (y) {
                  const { functionName: A, paramIndex: F } = y, N = r.getFunctionInfo(A);
                  c(N), g(F);
                  return;
                }
              }
          }
        }
        c(void 0), g(-1);
      }), p = n.selectionChange$.pipe(
        gr((S) => S.textRanges.length === 1),
        hr((S) => S.textRanges[0].startOffset),
        Yo()
      ).subscribe(() => {
        h(!0);
      });
      return () => {
        v.unsubscribe(), p.unsubscribe();
      };
    }
  }, [n, e]), $(() => {
    e || a();
  }, [e]), {
    functionInfo: i,
    paramIndex: l,
    reset: a
  };
}, Gi = ({ onClick: e }) => /* @__PURE__ */ M(
  "div",
  {
    className: "univer-z-[15] univer-box-border univer-h-[18px] univer-cursor-pointer univer-overflow-visible univer-whitespace-nowrap univer-rounded-l univer-border univer-border-r-0 univer-border-gray-600 univer-bg-primary-600 univer-p-0.5 univer-text-xs univer-font-bold univer-leading-[13px] univer-text-white",
    onClick: e,
    children: "?"
  }
), en = ({ className: e, title: t, value: n }) => /* @__PURE__ */ W("div", { className: "univer-my-2", children: [
  /* @__PURE__ */ M(
    "div",
    {
      className: re("univer-mb-2 univer-text-sm univer-font-medium univer-text-gray-900 dark:!univer-text-white", e),
      children: t
    }
  ),
  /* @__PURE__ */ M(
    "div",
    {
      className: "univer-whitespace-pre-wrap univer-break-words univer-text-xs univer-text-gray-500",
      children: n
    }
  )
] }), Zi = (e) => {
  const { prefix: t, value: n, active: o, onClick: r } = e;
  return /* @__PURE__ */ W("div", { children: [
    /* @__PURE__ */ W("span", { children: [
      t,
      "("
    ] }),
    n && n.map((s, i) => /* @__PURE__ */ W("span", { children: [
      /* @__PURE__ */ M(
        "span",
        {
          className: o === i ? "univer-text-primary-500" : "",
          onClick: () => r(i),
          children: Nr(s)
        }
      ),
      i === n.length - 1 ? "" : ","
    ] }, s.name)),
    ")"
  ] });
}, Xn = () => {
};
function zi(e) {
  const { onParamsSwitch: t = Xn, onClose: n = Xn, isFocus: o, editor: r, formulaText: s } = e, { functionInfo: i, paramIndex: c, reset: l } = Yi(o, s, r), g = T(dn), m = !Se(g.helpFunctionVisible$), [h, d] = V(!0), u = T(dt), a = u.t("formula.prompt.required"), v = u.t("formula.prompt.optional"), p = r.getEditorId(), [S] = Pr(p, !!i, [i, c]);
  function C(f) {
    t && t(f);
  }
  const _ = te((f) => {
    g.helpFunctionVisible$.next(!f);
  }), E = () => {
    _(!0), n();
  };
  return i ? m ? /* @__PURE__ */ M(nn, { portal: !0, anchorRect$: S, direction: "left-center", children: /* @__PURE__ */ M(Gi, { onClick: () => _(!1) }) }, "hidden") : /* @__PURE__ */ M(nn, { portal: !0, onClickOutside: () => l(), anchorRect$: S, direction: "vertical", children: /* @__PURE__ */ W(
    "div",
    {
      className: re("univer-m-0 univer-box-border univer-w-[250px] univer-select-none univer-list-none univer-rounded-lg univer-bg-white univer-leading-5 univer-shadow-md univer-outline-none dark:!univer-bg-gray-900", Sn),
      children: [
        /* @__PURE__ */ W(
          "div",
          {
            className: re("univer-wrap-anywhere univer-box-border univer-flex univer-items-center univer-justify-between univer-px-4 univer-py-3 univer-text-xs univer-font-medium univer-text-gray-900 dark:!univer-text-white", Es),
            children: [
              /* @__PURE__ */ M(
                Zi,
                {
                  prefix: i.functionName,
                  value: i.functionParameter,
                  active: c,
                  onClick: C
                }
              ),
              /* @__PURE__ */ W("div", { className: "univer-flex", children: [
                /* @__PURE__ */ M(
                  "div",
                  {
                    className: "univer-ml-2 univer-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-bg-transparent univer-p-0 univer-text-xs univer-text-gray-500 univer-outline-none univer-transition-colors hover:univer-bg-gray-200 dark:hover:!univer-bg-gray-600",
                    style: { transform: h ? "rotateZ(-90deg)" : "rotateZ(90deg)" },
                    onClick: () => d(!h),
                    children: /* @__PURE__ */ M(Fr, {})
                  }
                ),
                /* @__PURE__ */ M(
                  "div",
                  {
                    className: "univer-ml-2 univer-flex univer-h-6 univer-w-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-bg-transparent univer-p-0 univer-text-xs univer-text-gray-600 univer-outline-none univer-transition-colors hover:univer-bg-gray-300 dark:!univer-text-gray-200 dark:hover:!univer-bg-gray-600",
                    onClick: E,
                    children: /* @__PURE__ */ M(wr, {})
                  }
                )
              ] })
            ]
          }
        ),
        /* @__PURE__ */ M(
          "div",
          {
            className: re("univer-box-border univer-max-h-[350px] univer-overflow-y-auto univer-px-4 univer-pb-3 univer-pt-0", lt),
            style: {
              height: h ? "unset" : 0,
              padding: h ? "revert-layer" : 0
            },
            children: /* @__PURE__ */ W("div", { className: "univer-mt-3", children: [
              /* @__PURE__ */ M(
                en,
                {
                  title: u.t("formula.prompt.helpExample"),
                  value: `${i.functionName}(${i.functionParameter.map((f) => f.example).join(",")})`
                }
              ),
              /* @__PURE__ */ M(
                en,
                {
                  title: u.t("formula.prompt.helpAbstract"),
                  value: i.description
                }
              ),
              i && i.functionParameter && i.functionParameter.map((f, b) => /* @__PURE__ */ M(
                en,
                {
                  className: c === b ? "univer-text-primary-500" : "",
                  title: f.name,
                  value: `${f.require ? a : v} ${f.detail}`
                },
                b
              ))
            ] })
          }
        )
      ]
    }
  ) }, "show") : null;
}
const Xi = (e) => {
  const t = T(Ye);
  return te((o) => {
    var r, s;
    if (e) {
      t.focus(e.getEditorId());
      const i = [...e.getSelectionRanges()];
      if (ln.isDefine(o))
        e.setSelectionRanges([{ startOffset: o, endOffset: o }]);
      else if (!i.length && !e.docSelectionRenderService.isOnPointerEvent) {
        const c = (s = (r = e.getDocumentData().body) == null ? void 0 : r.dataStream) != null ? s : `\r
`, l = Math.max(c.length - 2, 0);
        e.setSelectionRanges([{ startOffset: l, endOffset: l }]);
      } else
        e.setSelectionRanges(i);
    }
  });
};
function Qi(e) {
  var r, s;
  const n = e.get(ne).getCurrentUniverDocInstance();
  return n != null && n.getBody() ? { dataStream: (s = (r = n.getBody()) == null ? void 0 : r.dataStream) != null ? s : "", offset: 0 } : void 0;
}
var Te = /* @__PURE__ */ ((e) => (e[e.NOT_SELECT = 0] = "NOT_SELECT", e[e.NEED_ADD = 1] = "NEED_ADD", e[e.CAN_EDIT = 2] = "CAN_EDIT", e[e.EDIT_OTHER_SHEET_REFERENCE = 3] = "EDIT_OTHER_SHEET_REFERENCE", e[e.EDIT_OTHER_WORKBOOK_REFERENCE = 4] = "EDIT_OTHER_WORKBOOK_REFERENCE", e))(Te || {});
function Ji(e) {
  var y;
  const { editorId: t, isFocus: n, disableOnClick: o, unitId: r, subUnitId: s } = e, i = T(xe), c = T(ne), l = i.getRenderById(r), g = i.getRenderById(t), m = g == null ? void 0 : g.with(mr), h = T(Rr), d = T(Rt), [u, a] = V(
    0
    /* NOT_SELECT */
  ), v = T(_e), p = z(!0), S = l == null ? void 0 : l.with(De), C = tt(u), _ = c.getUnit(r, H.UNIVER_SHEET), E = _ == null ? void 0 : _.getSheetBySheetId(s), f = te((A) => {
    S && S.setSkipLastEnabled(
      A === 1 || A === 3 || A === 4
      /* EDIT_OTHER_WORKBOOK_REFERENCE */
    ), C.current = A, a(A);
  }), b = te(() => {
    var ce, ae, he;
    const A = c.getCurrentUnitOfType(H.UNIVER_SHEET);
    if (!A) return;
    const F = A.getActiveSheet(), N = m == null ? void 0 : m.getActiveTextRange(), O = N != null && N.collapsed ? N.startOffset : -1, k = Qi(d);
    if (!k) return;
    const I = (ce = k == null ? void 0 : k.dataStream) == null ? void 0 : ce.slice(0, -2), x = ((ae = v.sequenceNodesBuilder(I)) != null ? ae : []).map((K) => typeof K == "object" ? K.nodeType === Q.REFERENCE ? {
      ...K,
      range: Lo(K.token)
    } : {
      ...K,
      range: void 0
    } : K), R = I[O - 1], w = I[O], D = x.find((K) => typeof K == "object" && K.nodeType === Q.REFERENCE && O === K.endIndex + 2), U = R && $o(R) && (!w || Po(w) && w !== je.OPEN_BRACKET), B = !!D;
    if ((I == null ? void 0 : I.substring(0, 1)) === "=" && (U || B))
      if (B) {
        if (p.current)
          return;
        const { sheetName: K, unitId: J } = D.range, nt = (he = c.getCurrentUnitOfType(H.UNIVER_SHEET)) == null ? void 0 : he.getUnitId();
        J && J !== nt ? f(
          4
          /* EDIT_OTHER_WORKBOOK_REFERENCE */
        ) : !K && F.getSheetId() === (E == null ? void 0 : E.getSheetId()) || K === F.getName() ? f(
          2
          /* CAN_EDIT */
        ) : f(
          3
          /* EDIT_OTHER_SHEET_REFERENCE */
        );
      } else
        p.current = !1, f(
          1
          /* NEED_ADD */
        );
    else
      f(
        0
        /* NOT_SELECT */
      );
  });
  return $(() => {
    const A = h.textSelection$.pipe(gr((F) => F.unitId === t)).subscribe(() => {
      b();
    });
    return () => A.unsubscribe();
  }, [b, h.textSelection$, t]), $(() => {
    n || (f(
      0
      /* NOT_SELECT */
    ), p.current = !0);
  }, [n, f]), $(() => {
    var F;
    if (!o) return;
    const A = (F = g == null ? void 0 : g.mainComponent) == null ? void 0 : F.onPointerDown$.subscribeEvent(() => {
      f(
        0
        /* NOT_SELECT */
      ), p.current = !0;
    });
    return () => A == null ? void 0 : A.unsubscribe();
  }, [o, (y = g == null ? void 0 : g.mainComponent) == null ? void 0 : y.onPointerDown$, f]), $(() => {
    if (!n) return;
    const A = _ == null ? void 0 : _.activeSheet$.subscribe(() => {
      b();
    }), F = c.getCurrentTypeOfUnit$(H.UNIVER_SHEET).subscribe(() => {
      b();
    });
    return () => {
      A == null || A.unsubscribe(), F == null || F.unsubscribe();
    };
  }, [b, n, _ == null ? void 0 : _.activeSheet$, c.getCurrentTypeOfUnit$]), { isSelecting: u, isSelectingRef: C };
}
const ec = () => {
  const e = T(_e);
  return Dt((n) => e.sequenceNodesBuilder(n) || [], [e]);
};
function tc(e, t, n) {
  const o = new un(t).setAlpha(0.05).toRgbString();
  return {
    id: n,
    strokeWidth: 1,
    stroke: t,
    fill: o,
    widgets: { tl: !0, tc: !0, tr: !0, ml: !0, mr: !0, bl: !0, bc: !0, br: !0 },
    widgetSize: 6,
    widgetStrokeWidth: 1,
    widgetStroke: e.getColorFromTheme("white")
  };
}
function Ur(e) {
  var _, E, f;
  const {
    unitId: t,
    subUnitId: n,
    currentWorkbook: o,
    refSelections: r,
    editor: s,
    refSelectionsService: i,
    refSelectionsRenderService: c,
    sheetSkeletonManagerService: l,
    themeService: g,
    univerInstanceService: m
  } = e, h = o.getUnitId(), d = m.getUnit(t, H.UNIVER_SHEET), u = d == null ? void 0 : d.getActiveSheet(), a = [];
  if (!d || !u) {
    i.setSelections(a);
    return;
  }
  const v = u.getSheetId(), p = (b) => {
    var y;
    return (y = d == null ? void 0 : d.getSheetBySheetName(b)) == null ? void 0 : y.getSheetId();
  };
  if (!((_ = l == null ? void 0 : l.getWorksheetSkeleton(v)) == null ? void 0 : _.skeleton)) return;
  const C = [];
  for (let b = 0, y = r.length; b < y; b++) {
    const A = r[b], { themeColor: F, token: N, refIndex: O, endIndex: k } = A, I = ht(N), { unitId: x, sheetName: R, range: w } = I, D = p(R);
    if (!D && R || h !== t && x !== h || x && x !== h || D && D !== v || !D && v !== n)
      continue;
    const U = as(w, u.getRowCount(), u.getColumnCount());
    U.unitId = t, U.sheetId = v, a.push({
      range: U,
      primary: null,
      style: tc(g, F, O.toString())
    }), C.push(k);
  }
  if (s) {
    const b = (f = (E = s.getSelectionRanges()) == null ? void 0 : E[0]) == null ? void 0 : f.startOffset, y = C.findIndex((A) => A + 2 === b);
    y !== -1 ? c == null || c.setActiveSelectionIndex(y) : c == null || c.resetActiveSelectionIndex();
  }
  return a;
}
function nc(e, t) {
  const n = T(ne), o = T(It), r = T(Kt), s = T(xe), i = Se(ie(() => n.getCurrentTypeOfUnit$(H.UNIVER_SHEET), [n])), c = i ? s.getRenderById(i.getUnitId()) : null, l = c == null ? void 0 : c.with(De), g = c == null ? void 0 : c.with(jt), m = te((h, d) => {
    const u = n.getCurrentUnitOfType(H.UNIVER_SHEET);
    if (!u || l != null && l.selectionMoving) return;
    const a = Ur({
      unitId: e,
      subUnitId: t,
      currentWorkbook: u,
      refSelections: h,
      editor: d,
      refSelectionsService: r,
      refSelectionsRenderService: l,
      sheetSkeletonManagerService: g,
      themeService: o,
      univerInstanceService: n
    });
    if (!a) return;
    ((l == null ? void 0 : l.getSelectionControls()) || []).length === a.length ? l == null || l.resetSelectionsByModelData(a) : r.setSelections(a);
  });
  return $(() => () => {
    l == null || l.resetActiveSelectionIndex();
  }, [l]), m;
}
function Vr(e = "") {
  const t = T(Yt), n = rc(), o = T(me), r = ie(() => e.length, [e]);
  return te((i, c, l = !0, g) => {
    const m = i.getDocumentData(), h = i.getEditorId();
    if (!m)
      return [];
    const d = m.body;
    if (!d)
      return [];
    const u = d.dataStream.slice(0, d.dataStream.length - 2), a = { dataStream: "", ...m.body };
    if (!u.startsWith(e)) return [];
    if (c == null || c.length === 0)
      return a.textRuns = [], o.syncExecuteCommand(Un.id, {
        unitId: h,
        body: Mn(a, 0, a.dataStream.length - 2)
      }), [];
    {
      const { textRuns: v, refSelections: p } = oc(t, n, c);
      r && v.forEach((_) => {
        _.ed = _.ed + r, _.st = _.st + r;
      }), a.textRuns = [{ st: 0, ed: 1, ts: { fs: 11 } }, ...v];
      const S = c.reduce((_, E) => typeof E == "string" ? `${_}${E}` : `${_}${E.token}`, "");
      a.dataStream = `${e}${S}\r
`;
      let C;
      if (l) {
        C = i.getSelectionRanges();
        const _ = a.dataStream.length - 2 + r;
        C.forEach((E) => {
          E.startOffset = Math.max(0, Math.min(E.startOffset, _)), E.endOffset = Math.max(0, Math.min(E.endOffset, _));
        });
      }
      return o.syncExecuteCommand(Un.id, {
        unitId: h,
        body: Mn(a, 0, a.dataStream.length - 2),
        textRanges: g != null ? g : C
      }), p;
    }
  });
}
function rc() {
  const e = T(It), t = e.getCurrentTheme();
  return ie(() => {
    const o = [
      e.getColorFromTheme("loop-color.1"),
      e.getColorFromTheme("loop-color.2"),
      e.getColorFromTheme("loop-color.3"),
      e.getColorFromTheme("loop-color.4"),
      e.getColorFromTheme("loop-color.5"),
      e.getColorFromTheme("loop-color.6"),
      e.getColorFromTheme("loop-color.7"),
      e.getColorFromTheme("loop-color.8"),
      e.getColorFromTheme("loop-color.9"),
      e.getColorFromTheme("loop-color.10"),
      e.getColorFromTheme("loop-color.11"),
      e.getColorFromTheme("loop-color.12")
    ].map((c) => e.isValidThemeColor(c) ? e.getColorFromTheme(c) : c), r = e.getColorFromTheme("blue.700"), s = e.getColorFromTheme("jiqing.800"), i = e.getColorFromTheme("black");
    return { formulaRefColors: o, numberColor: r, stringColor: s, plainTextColor: i };
  }, [t]);
}
function oc(e, t, n) {
  const { formulaRefColors: o, numberColor: r, stringColor: s, plainTextColor: i } = t, c = [], l = [], g = /* @__PURE__ */ new Map();
  let m = 0;
  for (let h = 0, d = n.length; h < d; h++) {
    const u = n[h];
    if (typeof u == "string") {
      const _ = c[c.length - 1], E = _ ? _.ed : 0, f = E + u.length;
      c.push({
        st: E,
        ed: f,
        ts: {
          cl: {
            rgb: i
          },
          fs: 11
        }
      });
      continue;
    }
    if (e.hasDefinedNameDescription(u.token.trim())) {
      c.push({
        st: u.startIndex,
        ed: u.endIndex + 1,
        ts: {
          cl: {
            rgb: i
          },
          fs: 11
        }
      });
      continue;
    }
    const { startIndex: a, endIndex: v, nodeType: p, token: S } = u;
    let C = "";
    if (p === Q.REFERENCE) {
      if (g.has(S))
        C = g.get(S);
      else {
        const _ = m % o.length;
        C = o[_], g.set(S, C), m++;
      }
      l.push({
        refIndex: h,
        themeColor: C,
        token: S,
        startIndex: u.startIndex,
        endIndex: u.endIndex,
        index: l.length
      });
    } else p === Q.NUMBER ? C = r : (p === Q.STRING || p === Q.ARRAY) && (C = s);
    C && C.length > 0 ? c.push({
      st: a,
      ed: v + 1,
      ts: {
        cl: {
          rgb: C
        },
        fs: 11
      }
    }) : c.push({
      st: a,
      ed: v + 1,
      ts: {
        cl: {
          rgb: i
        },
        fs: 11
      }
    });
  }
  return { textRuns: c, refSelections: l };
}
const sc = (e, t, n, o) => {
  const r = T(me), s = T(Gt), i = z(t);
  i.current = t;
  const c = z(o);
  c.current = o, $(() => {
    if (!n || !e)
      return;
    const g = `sheet.formula-embedding-editor.${n.getEditorId()}`, m = new qe(), h = (a, v) => {
      if (c.current) {
        c.current(a, v);
        return;
      }
      let p = de.LEFT;
      a === L.ARROW_DOWN ? p = de.DOWN : a === L.ARROW_UP ? p = de.UP : a === L.ARROW_RIGHT && (p = de.RIGHT), v === P.SHIFT ? r.executeCommand(Zo.id, {
        direction: p
      }) : r.executeCommand(zo.id, {
        direction: p
      });
    }, d = (a, v) => {
      let p = de.DOWN;
      a === L.ARROW_DOWN ? p = de.DOWN : a === L.ARROW_UP ? p = de.UP : a === L.ARROW_LEFT ? p = de.LEFT : a === L.ARROW_RIGHT && (p = de.RIGHT), i.current ? v === P.CTRL_COMMAND ? r.executeCommand(Dn.id, {
        direction: p,
        jumpOver: Ln.moveGap,
        extra: "formula-editor",
        fromCurrentSelection: i.current === Te.NEED_ADD || i.current === Te.EDIT_OTHER_SHEET_REFERENCE
      }) : v === P.SHIFT ? r.executeCommand($n.id, {
        direction: p,
        extra: "formula-editor"
      }) : v === (P.CTRL_COMMAND | P.SHIFT) ? r.executeCommand($n.id, {
        direction: p,
        jumpOver: Ln.moveGap,
        extra: "formula-editor"
      }) : r.executeCommand(Dn.id, {
        direction: p,
        extra: "formula-editor",
        fromCurrentSelection: i.current === Te.NEED_ADD || i.current === Te.EDIT_OTHER_SHEET_REFERENCE
      }) : h(a, v);
    };
    return m.add(r.registerCommand({
      id: g,
      type: Ne.OPERATION,
      handler(a, v) {
        const { keyCode: p, metaKey: S } = v;
        d(p, S);
      }
    })), [
      { keyCode: L.ARROW_DOWN },
      { keyCode: L.ARROW_LEFT },
      { keyCode: L.ARROW_RIGHT },
      { keyCode: L.ARROW_UP },
      { keyCode: L.ARROW_DOWN, metaKey: P.SHIFT },
      { keyCode: L.ARROW_LEFT, metaKey: P.SHIFT },
      { keyCode: L.ARROW_RIGHT, metaKey: P.SHIFT },
      { keyCode: L.ARROW_UP, metaKey: P.SHIFT },
      { keyCode: L.ARROW_DOWN, metaKey: P.CTRL_COMMAND },
      { keyCode: L.ARROW_LEFT, metaKey: P.CTRL_COMMAND },
      { keyCode: L.ARROW_RIGHT, metaKey: P.CTRL_COMMAND },
      { keyCode: L.ARROW_UP, metaKey: P.CTRL_COMMAND },
      { keyCode: L.ARROW_DOWN, metaKey: P.CTRL_COMMAND | P.SHIFT },
      { keyCode: L.ARROW_LEFT, metaKey: P.CTRL_COMMAND | P.SHIFT },
      { keyCode: L.ARROW_RIGHT, metaKey: P.CTRL_COMMAND | P.SHIFT },
      { keyCode: L.ARROW_UP, metaKey: P.CTRL_COMMAND | P.SHIFT }
    ].map(({ keyCode: a, metaKey: v }) => ({
      id: g,
      binding: v ? a | v : a,
      preconditions: () => !0,
      priority: 900,
      staticParameters: {
        eventType: Oe.Keyboard,
        keyCode: a,
        metaKey: v
      }
    })).forEach((a) => {
      m.add(s.registerShortcut(a));
    }), () => {
      m.dispose();
    };
  }, [r, n, e, s]);
}, ic = (e, t, n, o, r = !0) => {
  var u;
  const s = T(xe), i = T(cn), c = T(ps), l = T(Kt), g = T(ne), m = Se(ie(() => g.getCurrentTypeOfUnit$(H.UNIVER_SHEET), [g])), h = s.getRenderById((u = m == null ? void 0 : m.getUnitId()) != null ? u : ""), d = h == null ? void 0 : h.with(De);
  Lt(() => {
    if (e)
      return i.setContextValue(wn, !0), r && c.disable(), () => {
        const a = g.getCurrentUnitOfType(H.UNIVER_DOC);
        (a == null ? void 0 : a.getUnitId()) === o && i.setContextValue(wn, !1), r && c.enable(), l.clear();
      };
  }, [i, e, l, r, o]), Lt(() => {
    if (e && t) {
      const a = d == null ? void 0 : d.enableSelectionChanging();
      return i.setContextValue(Hn, !0), () => {
        i.setContextValue(Hn, !1), a == null || a.dispose();
      };
    }
  }, [i, e, d, t]), $(() => {
    e && (d == null || d.setSkipLastEnabled(!1));
  }, [e, d]);
}, cc = (e, t, n) => {
  const o = T(ne), r = T(fn);
  return Dt(() => {
    if (e) {
      const i = [...r.getWorkbookSelections(t).getSelectionsOfWorksheet(n)], c = o.getCurrentUnitForType(H.UNIVER_SHEET), l = c == null ? void 0 : c.getActiveSheet();
      (c == null ? void 0 : c.getUnitId()) !== t && o.setCurrentUnitForType(t), l && l.getSheetId() === n && r.setSelections(i);
    }
  }, [e, r, n, t, o]);
}, ac = (e) => e.reduce((t, n) => typeof n == "string" ? t + n.length : t + n.token.length, 0), sn = (e) => e.map((t) => typeof t == "string" ? t : t.token).join(""), kt = (e, t = !1, n = "", o = !1) => !t && !o ? e.map((r) => Ce(r.range)) : e.map((r) => o ? Uo(r) : r.sheetName !== "" && r.sheetName !== n ? Ct(r.sheetName, r.range) : Ce(r.range)), lc = (e) => {
  var m, h, d;
  const { editor: t, lexerTreeBuilder: n } = e, o = t == null ? void 0 : t.getSelectionRanges();
  if ((o == null ? void 0 : o.length) !== 1)
    return;
  const s = o[0].startOffset - 1, i = ((h = (m = t == null ? void 0 : t.getDocumentData().body) == null ? void 0 : m.dataStream) != null ? h : `\r
`).slice(0, -2), c = (d = n.sequenceNodesBuilder(i.slice(1))) != null ? d : [], l = _n(c, s, !1), g = $r(c, l);
  return {
    nodeIndex: l,
    updatingRefIndex: g,
    sequenceNodes: c,
    offset: s
  };
}, uc = (() => {
}), dc = (e, t, n, o, r, s, i, c, l, g = uc) => {
  var O;
  const m = T(xe), h = T(ne), d = T(me), u = T(Rr), a = T(It), v = T(_e), p = h.getUnit(o), S = te((k, I) => {
    var x, R, w;
    return (w = (R = (x = h.getUnit(k)) == null ? void 0 : x.getSheetBySheetId(I)) == null ? void 0 : R.getName()) != null ? w : "";
  }), C = ie(() => S(o, r), [S, r, o]), _ = Se(p == null ? void 0 : p.activeSheet$), E = tt({ activeSheet: _, sheetName: C }), f = Se(ie(() => h.getCurrentTypeOfUnit$(H.UNIVER_SHEET), [h])), b = m.getRenderById((O = f == null ? void 0 : f.getUnitId()) != null ? O : ""), y = b == null ? void 0 : b.with(De), A = b == null ? void 0 : b.with(jt), F = T(Kt), N = te((k, I) => {
    var B, ce, ae, he, K, J, nt, rt, yt, Re;
    const x = lc({ editor: l, lexerTreeBuilder: v });
    if (!x) return;
    const { nodeIndex: R, updatingRefIndex: w, sequenceNodes: D, offset: U } = x;
    if (n.current === Te.NEED_ADD)
      if (U !== 0) {
        if (R === -1 && D.length)
          return;
        const Y = k[k.length - 1], X = D.splice(R + 1), Ie = (B = Y.sheetId) != null ? B : r, fe = {
          range: Y,
          unitId: (ce = Y.unitId) != null ? ce : f.getUnitId(),
          sheetName: S((ae = Y.unitId) != null ? ae : f.getUnitId(), Ie)
        }, ee = Ie !== r, le = (f == null ? void 0 : f.getUnitId()) !== o, Ee = kt([fe], i && (ee || le), C, le);
        D.push({ token: Ee[0], nodeType: Q.REFERENCE });
        const $e = [...D, ...X], Me = sn($e);
        g(Me, ac(D), I);
      } else {
        const Y = k[k.length - 1], X = (he = Y.sheetId) != null ? he : r, Ie = {
          range: Y,
          unitId: (K = Y.unitId) != null ? K : f.getUnitId(),
          sheetName: S((J = Y.unitId) != null ? J : f.getUnitId(), X)
        }, fe = X !== r, ee = (f == null ? void 0 : f.getUnitId()) !== o, le = kt([Ie], i && (fe || ee), C, ee);
        D.unshift({ token: le[0], nodeType: Q.REFERENCE });
        const Ee = sn(D);
        g(Ee, le[0].length, I);
      }
    else if (n.current === Te.EDIT_OTHER_SHEET_REFERENCE || n.current === Te.EDIT_OTHER_WORKBOOK_REFERENCE) {
      const Y = k.pop();
      if (!Y) return;
      const X = D[R];
      if (typeof X == "object" && X.nodeType === Q.REFERENCE) {
        const Ie = X.token;
        (f == null ? void 0 : f.getUnitId()) !== o ? X.token = Vo((nt = f == null ? void 0 : f.getUnitId()) != null ? nt : "", C, Y) : X.token = C === (_ == null ? void 0 : _.getName()) ? Ce(Y) : Ct(_.getName(), Y);
        const ee = U + (X.token.length - Ie.length);
        g(Wo(D), ee, I);
      }
    } else {
      const Y = [...k];
      if (w !== -1) {
        const G = Y.pop();
        G && Y.splice(w, 0, G);
      }
      let X = 0;
      const Ie = D.map((G) => {
        var ge, ot, Pe, Ue;
        if (typeof G == "string")
          return G;
        if (G.nodeType === Q.REFERENCE) {
          const Ve = ht(G.token);
          if (Ve.sheetName || (Ve.sheetName = C), (Ve.unitId || o) !== (f == null ? void 0 : f.getUnitId()) || i && ((ge = E.current.activeSheet) == null ? void 0 : ge.getName()) !== Ve.sheetName)
            return G.token;
          const oe = Y[X];
          if (X++, !oe)
            return "";
          const be = (ot = oe.sheetId) != null ? ot : r, Tt = {
            range: oe,
            unitId: (Pe = oe.unitId) != null ? Pe : f.getUnitId(),
            sheetName: S((Ue = oe.unitId) != null ? Ue : f.getUnitId(), be)
          }, Nt = (f == null ? void 0 : f.getUnitId()) !== o;
          return kt([Tt], i && (be !== r || Nt), C, Nt)[0];
        }
        return G.token;
      });
      let fe = "", ee;
      Ie.forEach((G, ge) => {
        fe += G, ge === R && (ee = fe.length);
      });
      const le = [];
      for (let G = X; G <= k.length - 1; G++) {
        const ge = k[G], ot = (rt = ge.sheetId) != null ? rt : r, Pe = {
          range: ge,
          unitId: (yt = ge.unitId) != null ? yt : f.getUnitId(),
          sheetName: S((Re = ge.unitId) != null ? Re : f.getUnitId(), ot)
        }, Ue = (f == null ? void 0 : f.getUnitId()) !== o, oe = kt([Pe], i && (ot !== r || Ue), C, Ue);
        le.push(oe[0]);
      }
      const Ee = D[D.length - 1], $e = Ee && (typeof Ee == "string" ? !1 : Ee.nodeType === Q.REFERENCE), Me = `${fe}${le.length && $e ? "," : ""}${le.join(",")}`;
      g(Me, !le.length && ee ? ee : Me.length, I);
    }
  });
  $(() => {
    if (y && e) {
      let k = !0;
      const I = (R, w) => {
        if (k) {
          k = !1;
          return;
        }
        N(R.map((D) => D.rangeWithCoord), w);
      }, x = new qe();
      return x.add(y.selectionMoving$.subscribe((R) => {
        I(R, !1);
      })), x.add(y.selectionMoveEnd$.subscribe((R) => {
        I(R, !0);
      })), () => {
        x.dispose();
      };
    }
  }, [e, N, y]), $(() => {
    if (t && y && l) {
      const k = new qe(), I = () => {
        k.dispose(), y.getSelectionControls().forEach((w, D) => {
          k.add(
            w.selectionScaling$.subscribe((U) => {
              const B = y.getSelectionDataWithStyle().map((ae) => ae.rangeWithCoord), ce = B[D];
              U.sheetId = ce.sheetId, U.unitId = ce.unitId, B[D] = U, N(B, !1);
            })
          ), k.add(
            w.selectionMoving$.subscribe((U) => {
              const B = y.getSelectionDataWithStyle().map((ae) => ae.rangeWithCoord), ce = B[D];
              U.sheetId = ce.sheetId, U.unitId = ce.unitId, B[D] = U, N(B, !0);
            })
          );
        });
      }, x = Go(
        l.input$,
        F.selectionSet$,
        y.selectionMoveEnd$
      ).pipe(
        Ts(50)
      ).subscribe(() => {
        I();
      });
      return () => {
        x.unsubscribe(), k.dispose();
      };
    }
  }, [l, t, N, y, F.selectionSet$]), y == null || y.getSelectionDataWithStyle(), $(() => {
    if (c) {
      const k = d.onCommandExecuted((I) => {
        var R;
        if (I.id !== gn.id)
          return;
        const x = I.params;
        if (x.extra === "formula-editor" && x.selections.length) {
          const w = x.selections[x.selections.length - 1];
          if (w) {
            const D = n.current === Te.NEED_ADD, U = ((R = y == null ? void 0 : y.getSelectionDataWithStyle()) != null ? R : []).map((B) => B.rangeWithCoord);
            D ? U.push(w.range) : U[U.length - 1] = w.range, N(U, !0);
          }
        }
      });
      return () => {
        k.dispose();
      };
    }
  }, [d, l, n, v, c, N, y]), $(() => {
    if (!l)
      return;
    const k = u.textSelection$.subscribe((I) => {
      I.unitId === l.getEditorId() && Ur({
        unitId: o,
        subUnitId: r,
        refSelections: s.current,
        editor: l,
        refSelectionsService: F,
        refSelectionsRenderService: y,
        sheetSkeletonManagerService: A,
        themeService: a,
        univerInstanceService: h,
        currentWorkbook: f
      });
    });
    return () => k.unsubscribe();
  }, [u.textSelection$, l, s, y, F, A, r, a, o, h]);
}, hc = (e, t, n, o, r, s) => {
  const i = T(me), c = T(Ye), g = T(xe).getRenderById(t), m = T(ne), h = g == null ? void 0 : g.with(De);
  $(() => {
    if (e && h)
      if (n) {
        const d = () => {
          const v = h.getSelectionControls().length;
          for (let p = 1; p <= v; p++)
            h.clearLastSelection();
          return setTimeout(() => {
            s();
          }, 30);
        }, u = i.onCommandExecuted((v) => {
          v.id === Bn.id && d();
        }), a = m.getCurrentTypeOfUnit$(H.UNIVER_SHEET).subscribe((v) => {
          d();
        });
        return () => {
          u.dispose(), a.unsubscribe();
        };
      } else {
        const d = i.beforeCommandExecuted((u) => {
          if (u.id === Bn.id) {
            o(!1), r(), s();
            const a = c.getEditor(an);
            a == null || a.focus();
          }
        });
        return () => {
          d.dispose();
        };
      }
  }, [e, h]);
}, fc = (e, t, n) => {
  const o = T(_e), r = z(!0);
  $(() => {
    if (e) {
      const s = setTimeout(() => {
        r.current = !1;
      }, 500);
      return () => {
        clearTimeout(s);
      };
    }
  }, [e]), $(() => {
    if (!r.current && t) {
      const s = o.checkIfAddBracket(n);
      t(s === 0 && n.startsWith(dr.EQUALS), `${n}`);
    }
  }, [n, t]);
}, gc = (e, t = [], n) => {
  const o = T(Yt), [r, s] = V([]), [i, c] = V(""), l = z(-1), g = tt({ nodes: t }), m = () => {
    s([]), c(""), l.current = -1;
  };
  return $(() => {
    if (n && e) {
      const d = n.input$.pipe(hn(300)).subscribe(() => {
        const u = n.getSelectionRanges();
        if (u.length === 1) {
          const a = g.current.nodes, v = u[0];
          if (v.collapsed) {
            const p = _n(a, v.startOffset - 1, !1);
            l.current = p;
            const S = a[p];
            if (S && typeof S != "string" && S.nodeType === Q.FUNCTION) {
              l.current = p;
              const C = S.token, _ = o.getSearchListByNameFirstLetter(C);
              s(_), c(C);
              return;
            }
          }
        }
        l.current = -1, c(""), s((a) => a != null && a.length ? [] : a);
      });
      return () => {
        d.unsubscribe();
      };
    }
  }, [n, e]), $(() => {
    e || m();
  }, [e]), {
    searchList: r,
    searchText: i,
    handlerFormulaReplace: (d, u) => {
      const a = [...g.current.nodes];
      if (l.current !== -1) {
        const v = a.splice(l.current + 1), p = a.pop() || "";
        let S = (typeof p == "string" ? p.length : p.token.length) - d.length;
        return a.push(d), v[0] !== je.OPEN_BRACKET && u !== ur.DefinedName && (a.push(je.OPEN_BRACKET), S--), { text: sn([...a, ...v]), offset: S };
      }
    },
    reset: m
  };
}, mc = () => {
}, pc = Le(Sc);
function Sc(e, t) {
  const { isFocus: n, sequenceNodes: o, onSelect: r, editor: s, onClose: i = mc } = e, c = s.getEditorId(), l = T(Gt), g = T(me), { searchList: m, searchText: h, handlerFormulaReplace: d, reset: u } = gc(n, o, s), a = ie(() => !!m.length, [m]), v = z(void 0), [p, S] = V(0), C = z(!1), [_] = Pr(c, a, [h, m]), E = tt({ searchList: m, active: p }), f = (N, O) => {
    const k = d(N, O);
    k && (u(), r(k));
  };
  function b(N) {
    C.current && S(N);
  }
  function y() {
    C.current && S(-1);
  }
  $(() => {
    if (!m.length)
      return;
    const N = `sheet.formula-embedding-editor.search_function.${c}`, O = new qe(), k = (I) => {
      const { searchList: x, active: R } = E.current;
      switch (I) {
        case L.ARROW_UP: {
          S((w) => {
            const D = Math.max(0, w - 1);
            return A(D), D;
          });
          break;
        }
        case L.ARROW_DOWN: {
          S((w) => {
            const D = Math.min(x.length - 1, w + 1);
            return A(D), D;
          });
          break;
        }
        case L.TAB:
        case L.ENTER: {
          const w = x[R];
          f(w.name, w.functionType);
          break;
        }
        case L.ESC: {
          u(), i();
          break;
        }
      }
    };
    return O.add(g.registerCommand({
      id: N,
      type: Ne.OPERATION,
      handler(I, x) {
        const { keyCode: R } = x;
        k(R);
      }
    })), [L.ARROW_UP, L.ARROW_DOWN, L.ENTER, L.ESC, L.TAB].map((I) => ({
      id: N,
      binding: I,
      preconditions: () => !0,
      priority: 1e3,
      staticParameters: {
        eventType: Oe.Keyboard,
        keyCode: I
      }
    })).forEach((I) => {
      O.add(l.registerShortcut(I));
    }), () => {
      O.dispose();
    };
  }, [m]);
  function A(N) {
    const O = v.current;
    if (!O) return;
    const k = O.children[N];
    if (!k) return;
    const x = O.getBoundingClientRect().top, R = O.offsetHeight, w = k.getBoundingClientRect(), D = w.top, U = w.height;
    if (D >= 0 && D > x && D - x + U <= R)
      return;
    const B = k.offsetTop - (R - U) / 2;
    O.scrollTo({
      top: B,
      behavior: "smooth"
    });
  }
  const F = ie(() => {
    let N = "";
    return () => {
      clearTimeout(N), C.current = !0, N = setTimeout(() => {
        C.current = !1;
      }, 300);
    };
  }, []);
  return m.length > 0 && a && /* @__PURE__ */ M(nn, { portal: !0, anchorRect$: _, direction: "vertical", children: /* @__PURE__ */ M(
    "ul",
    {
      ref: (N) => {
        v.current = N, t && (t.current = N);
      },
      "data-u-comp": "sheets-formula-editor",
      className: re("univer-m-0 univer-box-border univer-max-h-[400px] univer-w-[250px] univer-list-none univer-overflow-y-auto univer-rounded-lg univer-bg-white univer-p-2 univer-leading-5 univer-shadow-md univer-outline-none dark:!univer-bg-gray-900", Sn, lt),
      children: m.map((N, O) => /* @__PURE__ */ W(
        "li",
        {
          className: re("univer-box-border univer-cursor-pointer univer-rounded univer-px-2 univer-py-1 univer-text-gray-900 univer-transition-colors dark:!univer-text-white", {
            "univer-bg-gray-200 dark:!univer-bg-gray-600": p === O
          }),
          onMouseEnter: () => b(O),
          onMouseLeave: y,
          onMouseMove: F,
          onClick: () => {
            f(N.name, N.functionType), s && s.focus();
          },
          children: [
            /* @__PURE__ */ W("span", { className: "univer-block univer-truncate univer-text-xs", children: [
              /* @__PURE__ */ M("span", { className: "univer-text-red-500", children: N.name.substring(0, h.length) }),
              /* @__PURE__ */ M("span", { children: N.name.slice(h.length) })
            ] }),
            /* @__PURE__ */ M(
              "span",
              {
                className: "univer-block univer-text-xs univer-text-gray-400",
                children: N.desc
              }
            )
          ]
        },
        N.name
      ))
    }
  ) });
}
const vc = (e) => e.startsWith(dr.EQUALS) ? e.slice(1) : "", Qn = () => {
}, Cc = Le((e, t) => {
  var En, bn, yn, Tn;
  const {
    errorText: n,
    initValue: o,
    unitId: r,
    subUnitId: s,
    isFocus: i = !0,
    isSupportAcrossSheet: c = !1,
    onFocus: l = Qn,
    onBlur: g = Qn,
    onChange: m,
    onVerify: h,
    className: d,
    editorId: u,
    moveCursor: a = !0,
    onFormulaSelectingChange: v,
    keyboardEventConfig: p,
    onMoveInEditor: S,
    resetSelectionOnBlur: C = !0,
    autoScrollbar: _ = !0,
    isSingle: E = !0,
    disableSelectionOnClick: f = !1,
    autofocus: b = !0,
    disableContextMenu: y,
    style: A
  } = e, F = T(Ye), N = z(null), O = te(m);
  _s(t, () => ({
    isClickOutSide: (Z) => N.current ? !N.current.contains(Z.target) : !1
  }));
  const k = te(v), I = z(null), x = z(void 0), R = x.current, [w, D] = V(i), U = z(null), B = ie(() => u != null ? u : so(`${xo}-${qt(4)}`), []), ce = ie(() => n !== void 0, [n]), ae = T(ne), he = ae.getUnit(B);
  Se(he == null ? void 0 : he.change$);
  const K = ec(), J = io.transform.getPlainText((bn = (En = he == null ? void 0 : he.getBody()) == null ? void 0 : En.dataStream) != null ? bn : ""), nt = tt(J), rt = ie(() => vc(J), [J]), yt = ie(() => K(rt), [rt, K]), { isSelecting: Re, isSelectingRef: Y } = Ji({ unitId: r, subUnitId: s, editorId: B, isFocus: w, disableOnClick: f }), X = z(""), fe = T(xe).getRenderById(B), ee = fe == null ? void 0 : fe.with(mr), le = ee == null ? void 0 : ee.isFocusing, Ee = ie(() => ae.getCurrentTypeOfUnit$(H.UNIVER_DOC), [ae]), $e = Se(Ee), Me = ($e == null ? void 0 : $e.getUnitId()) === B, G = z([]), ge = Re, Pe = (Tn = (yn = T(or).getConfig(yr)) == null ? void 0 : yn.functionScreenTips) != null ? Tn : !0;
  Ss(() => {
    O(J);
  }, [J, O]);
  const Ue = Vr("="), Ve = nc(r, s), oe = te((Z, se = !0, we, ye) => {
    if (!x.current) return;
    X.current = Z;
    const Ot = Z[0] === "=" ? Z.slice(1) : "", ke = K(Ot), qr = ke.reduce((We, Mt) => typeof Mt == "object" ? `${We}${Mt.token}` : `${We}${Mt}`, ""), xt = Ue(
      x.current,
      qr === Ot ? ke : [],
      se,
      ye
    );
    if (G.current = xt, we) {
      const We = ye != null ? ye : R == null ? void 0 : R.getSelectionRanges();
      if ((We == null ? void 0 : We.length) !== 1)
        return;
      const jr = We[0].startOffset - 1, Kr = _n(ke, jr, !1), Nn = $r(ke, Kr);
      if (Nn >= 0) {
        const On = xt.splice(Nn, 1)[0];
        On && xt.push(On);
      }
      Ve(w ? xt : [], x.current);
    }
  });
  $(() => {
    w && oe(J, !1, !0);
  }, [w]), $(() => {
    if (w) {
      if (X.current === J) return;
      oe(J, !1, !0);
    }
  }, [J]), fc(w, h, J);
  const be = Xi(R), Tt = cc(w, r, s);
  $(() => {
    var Z;
    k(Re, (Z = ee == null ? void 0 : ee.isFocusing) != null ? Z : !0);
  }, [k, Re]), Xo(w, p, R), Lt(() => {
    let Z;
    if (U.current) {
      Z = F.register({
        autofocus: b,
        editorUnitId: B,
        initialSnapshot: {
          id: B,
          body: {
            dataStream: `${o}\r
`,
            textRuns: [],
            customBlocks: [],
            customDecorations: [],
            customRanges: []
          },
          documentStyle: {}
        }
      }, U.current);
      const se = F.getEditor(B);
      x.current = se, oe(o, !1, !0);
    }
    return () => {
      Z == null || Z.dispose();
    };
  }, []), Lt(() => {
    i ? (D(i), be()) : (C && (R == null || R.blur(), Tt()), D(i));
  }, [i, R, be, Tt, C]);
  const { checkScrollBar: Nt } = Qo(R, E, _);
  ic(w, !!(Re && Me), r, B, y), sc(!!(w && le && a), ge, R, S);
  const Rn = te((Z, se, we) => {
    if (!le)
      return;
    const ye = se !== -1 ? [{ startOffset: se + 1, endOffset: se + 1, collapsed: !0 }] : void 0;
    oe(`=${Z}`, !0, we, ye), we && (be(), se !== -1 && setTimeout(() => {
      const Ot = { startOffset: se + 1, endOffset: se + 1 }, ke = R == null ? void 0 : R.render.with(Jo);
      ke == null || ke.scrollToRange({ ...Ot, collapsed: !0 });
    }, 50), Nt());
  });
  dc(
    w && !!(Re && Me),
    w,
    Y,
    r,
    s,
    G,
    c,
    !!ge,
    R,
    Rn
  ), hc(w && !!(Re && Me), r, c, D, g, () => {
    oe(nt.current, !1, !0);
  });
  const In = (Z) => {
    if (Z) {
      const se = R == null ? void 0 : R.getSelectionRanges();
      if (se && se.length === 1) {
        const we = se[0];
        if (we.collapsed) {
          const ye = Z.offset;
          setTimeout(() => {
            R == null || R.setSelectionRanges([{ startOffset: we.startOffset - ye, endOffset: we.endOffset - ye }]);
          }, 30);
        }
      }
      be(), oe(`=${Z.text}`);
    }
  }, Br = () => {
    D(!0), l(), be();
  };
  return /* @__PURE__ */ W("div", { className: d, children: [
    /* @__PURE__ */ M(
      "div",
      {
        ref: N,
        className: re("univer-relative univer-box-border univer-flex univer-h-full univer-w-full univer-items-center univer-justify-around univer-gap-2 univer-rounded-none univer-p-0 univer-ring-1", {
          "univer-ring-primary-500": w,
          "univer-ring-red-500": ce
        }),
        children: /* @__PURE__ */ M(
          "div",
          {
            ref: U,
            className: "univer-relative univer-h-full univer-w-full",
            onMouseUp: Br
          }
        )
      }
    ),
    n !== void 0 && /* @__PURE__ */ M("div", { className: "univer-my-1 univer-text-xs univer-text-red-500", children: n }),
    Pe && R && rt !== "" && /* @__PURE__ */ M(
      zi,
      {
        editor: R,
        isFocus: w,
        formulaText: J,
        onClose: () => be()
      }
    ),
    Pe && !!R && /* @__PURE__ */ M(
      pc,
      {
        isFocus: w,
        sequenceNodes: yt,
        onSelect: In,
        ref: I,
        editor: R
      }
    )
  ] });
});
function _c(e, t, n, o) {
  const r = T(_e), s = Vr(""), i = Se(e == null ? void 0 : e.getDocumentDataModel().change$), [c, l] = V([]), g = T(Mo), m = z(""), h = T(ne);
  return $(() => {
    if (!e) return;
    const d = e.getDocumentDataModel().getPlainText();
    if (m.current === d)
      return;
    m.current = d;
    const u = r.sequenceNodesBuilder(d);
    l(u != null ? u : []);
  }, [i, e, r]), $(() => {
    var a, v;
    if (!e) return;
    if (!t) {
      const p = e.getDocumentData();
      e.setDocumentData({
        ...p,
        body: {
          ...p.body,
          dataStream: (v = (a = p.body) == null ? void 0 : a.dataStream) != null ? v : "",
          textRuns: []
        }
      });
      return;
    }
    const d = s(e, c, !1), u = new qe();
    return d.forEach((p) => {
      const S = ht(p.token), C = h.getCurrentUnitForType(H.UNIVER_SHEET), _ = C == null ? void 0 : C.getActiveSheet();
      if (!S.sheetName && o !== (_ == null ? void 0 : _.getSheetId()) || S.sheetName && (_ == null ? void 0 : _.getName()) !== S.sheetName)
        return;
      const E = new un(p.themeColor).toRgb(), f = g.addShape({
        range: S.range,
        style: {
          stroke: p.themeColor,
          fill: `rgba(${E.r}, ${E.g}, ${E.b}, 0.1)`,
          strokeDash: 12
        },
        primary: null
      });
      f && u.add(() => g.removeShape(f));
    }), () => {
      u.dispose();
    };
  }, [e, t, s, g, c]), { sequenceNodes: c };
}
function Rc(e) {
  const t = T(fn), { supportAcrossSheet: n = !1, keepSheetReference: o = !1, unitId: r, subUnitId: s, onChange: i } = e, l = T(ne).getUnit(r, H.UNIVER_SHEET), g = te(i), m = te((h, d) => {
    const u = l == null ? void 0 : l.getActiveSheet();
    if (!u || !n && u.getSheetId() !== s || !(h != null && h.length)) return;
    const a = o ? u.getName() : u.getSheetId() === s ? "" : u.getName(), v = h.map((p) => ({
      range: p.range,
      unitId: r,
      sheetName: a
    }));
    g(v, d);
  });
  $(() => {
    const h = new qe();
    return h.add(t.selectionMoveStart$.subscribe((d) => {
      m(d, !0);
    })), h.add(t.selectionMoving$.subscribe((d) => {
      m(d, !1);
    })), h.add(t.selectionMoveEnd$.subscribe((d) => {
      m(d, !1);
    })), () => {
      h.dispose();
    };
  }, [m, t.selectionMoveEnd$, t.selectionMoveStart$, t.selectionMoving$]);
}
const Jn = (e) => !e.some((n) => {
  if (typeof n == "string") {
    if (n !== je.COMMA)
      return !0;
  } else if (n.nodeType !== Q.REFERENCE)
    return !0;
  return !1;
}), Ic = (e) => {
  if (e.endColumn < e.startColumn) {
    const t = e.endColumn;
    e.endColumn = e.startColumn, e.startColumn = t;
  }
  if (e.endRow < e.startRow) {
    const t = e.endRow;
    e.endRow = e.startRow, e.startRow = t;
  }
  return e;
};
function Ec(e) {
  const {
    visible: t,
    initialValue: n,
    unitId: o,
    subUnitId: r,
    maxRangeCount: s = 1 / 0,
    supportAcrossSheet: i,
    keepSheetReference: c,
    onConfirm: l,
    onClose: g,
    onShowBySelection: m
  } = e, h = T(dt), d = T(_e), [u, a] = V([]), [v, p] = V(0), S = z(null);
  $(() => {
    if (t && n.length) {
      const f = n.map((b) => b.sheetName ? Ct(b.sheetName, b.range) : Ce(b.range));
      a(f), p(f.length - 1);
    } else
      a([""]), p(0);
  }, [t]);
  const C = (f, b) => {
    const y = [...u];
    y[f] = b, a(y);
  }, _ = () => {
    a([...u, ""]), p(u.length);
  }, E = (f) => {
    u.splice(f, 1), a([...u]);
  };
  return Rc({
    unitId: o,
    subUnitId: r,
    supportAcrossSheet: i,
    keepSheetReference: c,
    onChange: (f, b) => {
      if (!t && m != null && m(f))
        return;
      const y = new Set(u), A = f.map((O) => O.sheetName ? Ct(O.sheetName, O.range) : Ce(O.range)), F = A.filter((O) => !y.has(O));
      if (!F.length) return;
      const N = [...u];
      if (A.length > 1) {
        b || N.splice(v, 1), N.push(...F);
        const O = N.slice(0, s);
        a(O), p(O.length - 1), requestAnimationFrame(() => {
          var k;
          (k = S.current) == null || k.scrollTo({ top: S.current.scrollHeight });
        });
      } else {
        N.splice(v, 1, ...F);
        const O = N.slice(0, s);
        a(O), p(v + F.length - 1);
      }
    }
  }), /* @__PURE__ */ M(
    ys,
    {
      width: "328px",
      open: t,
      title: h.t("rangeSelector.title"),
      draggable: !0,
      mask: !1,
      maskClosable: !1,
      footer: /* @__PURE__ */ W("footer", { className: "univer-flex univer-gap-2", children: [
        /* @__PURE__ */ M(ct, { onClick: g, children: h.t("rangeSelector.cancel") }),
        /* @__PURE__ */ M(
          ct,
          {
            variant: "primary",
            onClick: () => {
              l(
                u.filter((f) => {
                  const b = d.sequenceNodesBuilder(f);
                  return b && b.length === 1 && typeof b[0] != "string" && b[0].nodeType === Q.REFERENCE;
                }).map((f) => ht(f)).map((f) => ({ ...f, range: Ic(f.range) }))
              );
            },
            children: h.t("rangeSelector.confirm")
          }
        )
      ] }),
      onClose: g,
      children: /* @__PURE__ */ W(
        "div",
        {
          ref: S,
          className: re("-univer-mx-6 univer-max-h-60 univer-overflow-y-auto univer-px-6", lt),
          children: [
            u.map((f, b) => /* @__PURE__ */ W(
              "div",
              {
                className: "univer-mb-2 univer-flex univer-items-center univer-gap-4",
                children: [
                  /* @__PURE__ */ M(
                    _r,
                    {
                      className: re("univer-w-full", {
                        "univer-border-primary-600": v === b
                      }),
                      placeholder: h.t("rangeSelector.placeHolder"),
                      onFocus: () => p(b),
                      value: f,
                      onChange: (y) => C(b, y)
                    }
                  ),
                  u.length > 1 && /* @__PURE__ */ M(
                    kr,
                    {
                      className: "univer-cursor-pointer",
                      onClick: () => E(b)
                    }
                  )
                ]
              },
              b
            )),
            u.length < s && /* @__PURE__ */ M("div", { children: /* @__PURE__ */ W(ct, { variant: "link", onClick: _, children: [
              /* @__PURE__ */ M(Ar, {}),
              /* @__PURE__ */ M("span", { children: h.t("rangeSelector.addAnotherRange") })
            ] }) })
          ]
        }
      )
    }
  );
}
function bc(e) {
  return e.split(je.COMMA).filter((t) => !!t).map((t) => ht(t));
}
function yc(e) {
  return e.map((t) => t.sheetName ? Ct(t.sheetName, t.range) : Ce(t.range)).join(je.COMMA);
}
function Wr(e) {
  const [t, n] = V(null), {
    onVerify: o,
    selectorRef: r,
    unitId: s,
    subUnitId: i,
    maxRangeCount: c,
    supportAcrossSheet: l,
    keepSheetReference: g,
    autoFocus: m,
    onChange: h,
    onRangeSelectorDialogVisibleChange: d,
    onClickOutside: u,
    onFocusChange: a,
    forceShowDialogWhenSelectionChanged: v,
    hideEditor: p,
    resetRange: S
  } = e, [C, _] = V(m != null ? m : !1), [E, f] = V(!1), [b, y] = V([]), A = T(dt), F = T(Ye), { sequenceNodes: N } = _c(t, C, s, i), O = tt(N), k = T(me), I = te(() => {
    t == null || t.setSelectionRanges([]), t == null || t.blur(), F.blur();
  }), x = te(() => {
    var R;
    I(), y(bc((R = t == null ? void 0 : t.getDocumentDataModel().getPlainText()) != null ? R : "")), f(!0);
  });
  return $(() => {
    r && (r.current = {
      get editor() {
        return t;
      },
      focus() {
        F.focus(t.getEditorId());
      },
      blur: I,
      verify: () => Jn(O.current),
      showDialog: (R) => {
        I(), y(R), f(!0);
      },
      hideDialog: () => {
        y([]), f(!1);
      },
      getValue: () => {
        var R;
        return (R = t == null ? void 0 : t.getDocumentDataModel().getPlainText()) != null ? R : "";
      }
    });
  }, [I, t, F, r, O]), $(() => {
    var R;
    o == null || o(Jn(N), (R = t == null ? void 0 : t.getDocumentDataModel().getPlainText()) != null ? R : "");
  }, [N]), $(() => {
    d == null || d(E);
  }, [E]), $(() => {
    if (E && S)
      return () => {
        const R = {
          unitId: s,
          subUnitId: i,
          selections: S
        };
        k.executeCommand(gn.id, R);
      };
  }, [E]), /* @__PURE__ */ W(Cs, { children: [
    p ? null : /* @__PURE__ */ M(
      es,
      {
        isSingle: !0,
        ...e,
        onFocusChange: (R, w) => {
          _(R), a == null || a(R, w);
        },
        editorRef: n,
        onClickOutside: () => {
          _(!1), I(), u == null || u();
        },
        icon: /* @__PURE__ */ M(bs, { title: A.t("rangeSelector.buttonTooltip"), placement: "bottom", children: /* @__PURE__ */ M(
          Dr,
          {
            className: "univer-cursor-pointer dark:!univer-text-gray-300",
            onClick: x
          }
        ) })
      }
    ),
    /* @__PURE__ */ M(
      Ec,
      {
        initialValue: b,
        unitId: s,
        subUnitId: i,
        visible: E,
        maxRangeCount: c,
        onConfirm: (R) => {
          const w = yc(R), D = co.newEmptyData();
          D.body.dataStream = w, t == null || t.replaceText(w, !1), h == null || h(D, w), f(!1), y([]), requestAnimationFrame(() => {
            I();
          });
        },
        onClose: () => {
          f(!1), y([]);
        },
        supportAcrossSheet: l,
        keepSheetReference: g,
        onShowBySelection: (R) => C || v ? (y(R), f(!0), !1) : !0
      }
    )
  ] });
}
const Tc = () => {
  var o, r;
  const e = T(Lr), t = Se(e.currentSelector$), n = z(null);
  return $(() => {
    var s, i;
    if (t)
      return (i = n.current) == null || i.showDialog((s = t.initialValue) != null ? s : []), () => {
        var c;
        (c = n.current) == null || c.hideDialog();
      };
  }, [t]), /* @__PURE__ */ M(
    Wr,
    {
      unitId: (o = t == null ? void 0 : t.unitId) != null ? o : "",
      subUnitId: (r = t == null ? void 0 : t.subUnitId) != null ? r : "",
      hideEditor: !0,
      selectorRef: n,
      onChange: (s, i) => {
        var c;
        t == null || t.callback((c = i == null ? void 0 : i.split(",").map((l) => ht(l))) != null ? c : []);
      }
    }
  );
};
var Nc = Object.defineProperty, Oc = Object.getOwnPropertyDescriptor, xc = (e, t, n) => t in e ? Nc(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, Mc = (e, t, n, o) => {
  for (var r = o > 1 ? void 0 : o ? Oc(t, n) : t, s = e.length - 1, i; s >= 0; s--)
    (i = e[s]) && (r = i(r) || r);
  return r;
}, At = (e, t) => (n, o) => t(n, o, e), Hr = (e, t, n) => xc(e, typeof t != "symbol" ? t + "" : t, n);
let Bt = class extends lo {
  constructor(e = Gn, t, n, o, r) {
    super(), this._config = e, this._injector = t, this._renderManagerService = n, this._configService = o, this._uiPartsService = r;
    const { menu: s, ...i } = uo(
      Gn,
      this._config
    );
    s && this._configService.setConfig("menu", s, { merge: !0 }), this._configService.setConfig(yr, i, { merge: !0 });
  }
  onStarting() {
    ho(this._injector, [
      [Zt, { useClass: rn }],
      [Lr],
      [Ht],
      [Pt],
      [Ut],
      [Vt],
      [Wt],
      [$t]
    ]), this._initUIPart();
  }
  onReady() {
    [
      [De]
    ].forEach((e) => {
      this.disposeWithMe(this._renderManagerService.registerRenderModule(H.UNIVER_SHEET, e));
    });
  }
  onRendered() {
    [
      [on]
    ].forEach((e) => {
      this.disposeWithMe(this._renderManagerService.registerRenderModule(H.UNIVER_SHEET, e));
    }), fo(this._injector, [
      [Ht],
      // FormulaProgressBar relies on TriggerCalculationController, but it is necessary to ensure that the formula calculation is done after rendered.
      [Ut],
      [Wt]
    ]);
  }
  onSteady() {
    this._injector.get(Pt), this._injector.get($t);
  }
  _initUIPart() {
    const e = this._injector.get(vr);
    this.disposeWithMe(e.register(wo, Wr)), this.disposeWithMe(e.register(ko, Cc)), this.disposeWithMe(this._uiPartsService.registerComponent(vs.GLOBAL, () => Cr(Tc, this._injector)));
  }
};
Hr(Bt, "pluginName", Ir);
Hr(Bt, "type", H.UNIVER_SHEET);
Bt = Mc([
  ao(Ho, hs),
  At(1, q(Rt)),
  At(2, xe),
  At(3, or),
  At(4, Sr)
], Bt);
export {
  Ms as FORMULA_PROMPT_ACTIVATED,
  Cc as FormulaEditor,
  $t as FormulaReorderController,
  Lr as GlobalRangeSelectorService,
  ws as HelpFunctionOperation,
  pe as InsertFunctionOperation,
  Cn as MoreFunctionsOperation,
  Wr as RangeSelector,
  De as RefSelectionsRenderService,
  br as ReferenceAbsoluteOperation,
  Ps as SearchFunctionOperation,
  gt as SelectEditorFormulaOperation,
  vn as SheetOnlyPasteFormulaCommand,
  Bt as UniverSheetsFormulaUIPlugin
};
